
# Appendix/Supplementary information

![
**GTDB species used in this paper.** These species were used to benchmark pangenome construction with reduced alphabet k-mers and open reading frame prediction from short sequencing reads. The trees are the default GTDB rs202 trees, with tips representing species not used in this paper removed.
](images/tree_fig.png "tree fig"){#fig:tree_fig tag="S1"}

![
**K-mer size and encoding do not impact pangenome estimation with k-mers.** Box plots representing the distribution of R^2^ values for linear models (Total, Unique) or statistic values for mantel tests (Mantel) calculated for each pangenome. All pangenomes are included, whether they contain genomes with the RefSeq exclusion criteria "many frameshifted proteins" or not. See figure legend for **Figure @fig:panmers_fig** for a description of Total, Unique, and Mantel.
](images/boxplt_fig.png){#fig:boxplt_fig tag="S2"}

![
**Percent of reads encoding coding domain sequences (CDS) that were predicted to be coding.** There is no change between the percent of reads predicted to be derived from coding domain sequences when a species-level database is used versus when all of GTDB is used to predict open reading frames
The slight increase observable for some species is a result of different thresholds, where we used 0.39 for the species database and 0.5 for the GTDB rs202 database.
](images/orpheum_supp1.png){#fig:orpheum_db  tag="S3" height=3in}


|species| n |total  |core |  shell  | cloud | alpha |
|:--------------------------:|:-:|:-----:|:-----:|:-----:|:-----:|:-----:|
|*Bacteroides fragilis*      |  7| 24819 |56.3%   | 11.3% | 32.4% |0.76|
|*Bacteroides uniformis*     |  9| 32197 |38.0%   | 22.3% | 39.7% |0.73|
|*Enterocloster bolteae*     |  4| 23620 |55.8%   | 18.3% | 25.9% |0.66|
|*Parabacteroides distasonis*|  7| 25789 |42.4%   | 30.9% | 26.8% |0.74|
|*Parabacteroides merdae*    |  6| 19985 |63.2%   | 9.6%  | 27.1% |0.82|
|*Phocaeicola vulgatus*      | 11| 41005 |30.3%   |20.4%  | 49.2% |0.65|
Table: Metapangenome estimates for each species. *n* designates the number of metagenomes used to estimate the total, core, shell, cloud, and alpha values. Unlike isolate genomes, metagenomes may contain a fraction of an organism's genome if the metagenome was not sequenced deeply or if an organism was rare. 
To calculate the core, shell, and cloud fractions and to estimate the openness of the metapangenome, we removed samples with fewer than 10,000 k~aa~-mers.
{#tbl:metap tag="S1"}

![**k~aa~-mer metapangenomes for six species.** Each species contains a four-panel figure.
The first panel is a binmap plot. 
Dark colors represent k-mers that are present in each sample. 
Blue shades represent time points when the sampled individual was not on antibiotics, while green shades represent time points when the individual was on antibiotics. 
The second panel represents an estimated number of base pairs in the metagenome detected to originate from that species. 
The third panel represents an estimated fraction of the metagenome assigned to that species.
The fourth panel represents the number of bins produced for that species from that sample using a *de novo* metagenome assembly and binning approach.
The two values represented in the second and third panels and the species assignations used to infer the value represented in the fourth panel were inferred using the sourmash gather algorithm against the GTDB rs202 database.
**A)** Species for which presence-absence fluctuated over the time series. 
**B)** Species for which strain presence-absence fluctuated over the time series.
](images/metap.png){#fig:metap_sfig tag="S4"}

| accession       | superkingdom | phylum                | class                  | order                 | family                   | genus                   | species                             | NCBI taxid | NCBI organism name                                       |
|-----------------|--------------|-----------------------|------------------------|-----------------------|--------------------------|-------------------------|-------------------------------------|------------|----------------------------------------------------------|
| GCF_000012125.1 | d__Bacteria  | p__Chlamydiota        | c__Chlamydiia          | o__Chlamydiales       | f__Chlamydiaceae         | g__Chlamydia            | s__Chlamydia trachomatis            | 315277     | Chlamydia trachomatis A/HAR-13                           |
| GCF_000381045.1 | d__Bacteria  | p__Deinococcota       | c__Deinococci          | o__Deinococcales      | f__Thermaceae            | g__Thermus              | s__Thermus scotoductus              | 1123391    | Thermus scotoductus DSM 8553                             |
| GCF_900156205.1 | d__Bacteria  | p__Spirochaetota      | c__Leptospirae         | o__Leptospirales      | f__Leptospiraceae        | g__Leptospira           | s__Leptospira interrogans           | 173        | Leptospira interrogans                                   |
| GCA_900315925.1 | d__Bacteria  | p__Firmicutes_C       | c__Negativicutes       | o__Acidaminococcales  | f__Acidaminococcaceae    | g__Succiniclasticum     | s__Succiniclasticum sp900315925     | 1387507    | uncultured Selenomonadales bacterium                     |
| GCF_004006635.1 | d__Bacteria  | p__Fusobacteriota     | c__Fusobacteriia       | o__Fusobacteriales    | f__Fusobacteriaceae      | g__Fusobacterium_C      | s__Fusobacterium_C necrophorum      | 143388     | Fusobacterium necrophorum subsp. necrophorum             |
| GCA_903857495.1 | d__Bacteria  | p__Acidobacteriota    | c__Holophagae          | o__Holophagales       | f__Holophagaceae         | g__Geothrix             | s__Geothrix sp903857495             | 904990     | uncultured Holophagaceae bacterium                       |
| GCF_006742205.1 | d__Bacteria  | p__Firmicutes         | c__Bacilli             | o__Staphylococcales   | f__Staphylococcaceae     | g__Staphylococcus       | s__Staphylococcus epidermidis       | 1282       | Staphylococcus epidermidis                               |
| GCF_002006445.1 | d__Bacteria  | p__Firmicutes_A       | c__Clostridia          | o__Clostridiales      | f__Clostridiaceae        | g__Clostridium          | s__Clostridium beijerinckii         | 1520       | Clostridium beijerinckii                                 |
| GCA_903861715.1 | d__Bacteria  | p__Patescibacteria    | c__Paceibacteria       | o__Moranbacterales    | f__GWC2-37-73            | g__CAIPMZ01             | s__CAIPMZ01 sp903861715             | 77133      | uncultured bacterium                                     |
| GCF_000008025.1 | d__Bacteria  | p__Proteobacteria     | c__Alphaproteobacteria | o__Rickettsiales      | f__Anaplasmataceae       | g__Wolbachia            | s__Wolbachia pipientis              | 163164     | Wolbachia endosymbiont of Drosophila melanogaster        |
| GCF_000830885.1 | d__Bacteria  | p__Chloroflexota      | c__Dehalococcoidia     | o__Dehalococcoidales  | f__Dehalococcoidaceae    | g__Dehalococcoides      | s__Dehalococcoides mccartyi_B       | 1432061    | Dehalococcoides mccartyi CG5                             |
| GCF_000299235.1 | d__Bacteria  | p__Nitrospirota_A     | c__Leptospirillia      | o__Leptospirillales   | f__Leptospirillaceae     | g__Leptospirillum_A     | s__Leptospirillum_A rubarum         | 1048260    | Leptospirillum ferriphilum ML-04                         |
| GCA_000635495.1 | d__Bacteria  | p__Cyanobacteria      | c__Cyanobacteriia      | o__PCC-6307           | f__Cyanobiaceae          | g__Prochlorococcus_A    | s__Prochlorococcus_A sp000635495    | 1471472    | Prochlorococcus sp. scB243_495K23                        |
| GCA_903852495.1 | d__Bacteria  | p__Desulfobacterota_F | c__Desulfuromonadia    | o__Geobacterales      | f__Pseudopelobacteraceae | g__Pseudopelobacter     | s__Pseudopelobacter sp903852495     | 214033     | uncultured Geobacteraceae bacterium                      |
| GCA_903869265.1 | d__Bacteria  | p__Desulfobacterota   | c__Desulfobulbia       | o__Desulfobulbales    | f__Desulfurivibrionaceae | g__UBA2262              | s__UBA2262 sp903869265              | 34034      | uncultured delta proteobacterium                         |
| GCF_000020225.1 | d__Bacteria  | p__Verrucomicrobiota  | c__Verrucomicrobiae    | o__Verrucomicrobiales | f__Akkermansiaceae       | g__Akkermansia          | s__Akkermansia muciniphila          | 349741     | Akkermansia muciniphila ATCC BAA-835                     |
| GCF_900478295.1 | d__Bacteria  | p__Campylobacterota   | c__Campylobacteria     | o__Campylobacterales  | f__Helicobacteraceae     | g__Helicobacter         | s__Helicobacter pylori              | 102618     | Helicobacter pylori NCTC 11637 = CCUG 17874 = ATCC 43504 |
| GCF_000154205.1 | d__Bacteria  | p__Bacteroidota       | c__Bacteroidia         | o__Bacteroidales      | f__Bacteroidaceae        | g__Bacteroides          | s__Bacteroides uniformis            | 411479     | Bacteroides uniformis ATCC 8492                          |
| GCF_000196555.1 | d__Bacteria  | p__Actinobacteriota   | c__Actinomycetia       | o__Actinomycetales    | f__Bifidobacteriaceae    | g__Bifidobacterium      | s__Bifidobacterium longum           | 565042     | Bifidobacterium longum subsp. longum JCM 1217            |
| GCF_000012285.1 | d__Archaea   | p__Thermoproteota     | c__Thermoproteia       | o__Sulfolobales       | f__Sulfolobaceae         | g__Sulfolobus           | s__Sulfolobus acidocaldarius        | 330779     | Sulfolobus acidocaldarius DSM 639                        |
| GCF_000970205.1 | d__Archaea   | p__Halobacteriota     | c__Methanosarcinia     | o__Methanosarcinales  | f__Methanosarcinaceae    | g__Methanosarcina       | s__Methanosarcina mazei             | 213585     | Methanosarcina mazei S-6                                 |
| GCA_900314695.1 | d__Archaea   | p__Methanobacteriota  | c__Methanobacteria     | o__Methanobacteriales | f__Methanobacteriaceae   | g__Methanobrevibacter_A | s__Methanobrevibacter_A sp900314695 | 253161     | uncultured Methanobrevibacter sp.                        |
| GCF_900083515.1 | d__Archaea   | p__Thermoplasmatota   | c__Thermoplasmata      | o__Thermoplasmatales  | f__Thermoplasmataceae    | g__Cuniculiplasma       | s__Cuniculiplasma divulgatum        | 1673428    | Cuniculiplasma divulgatum                                |
Table: GTDB genomes used to benchmark orpheum accuracy.
{#tbl:gtdb_acc tag="S2"}


| accession       | NCBI taxid | NCBI organism name                      |
|-----------------|------------|-----------------------------------------|
| GCF_003428625.2 | 2303751    | Acidipila sp. 4G-K13                    |
| GCF_001700755.2 | 1160719    | Cutibacterium granulosum DSM 20700      |
| GCF_001884725.2 | 336810     | Candidatus Sulcia muelleri              |
| GCF_015356815.2 | 225148     | Candidatus Rhabdochlamydia porcellionis |
| GCF_019599295.1 | 2866714    | Oscillochloris sp. ZM17-4               |
| GCF_020520145.1 | 936456     | Desulfurispirillum indicum              |
| GCF_019175305.1 | 2286       | Saccharolobus shibatae                  |
| GCF_018282115.1 | 2732530    | Synechocystis sp. PCC 7338              |
| GCF_018863415.1 | 2774531    | Deinococcus sp. SYSU M49105             |
| GCF_013456555.2 | 1710539    | Natrinema sp. YPL30                     |
| GCF_000167435.2 | 1314       | Streptococcus pyogenes                  |
| GCF_018205295.1 | 859        | Fusobacterium necrophorum               |
| GCF_019173545.1 | 1455061    | Candidatus Magnetobacterium casensis    |
| GCF_018398935.1 | 1123043    | Telmatocola sphagniphila                |
| GCF_000145825.2 | 629264     | Pseudomonas syringae Cit 7              |
| GCF_002442595.2 | 139        | Borreliella burgdorferi                 |
| GCF_009156025.2 | 28903      | Mycoplasmopsis bovis                    |
| GCF_019688735.1 | 2867247    | Thermosulfurimonas sp. F29              |
| GCF_018588215.1 | 1755816    | Thermosipho sp. 1244                    |
| GCF_018336995.1 | 239935     | Akkermansia muciniphila                 |
Table: RefSeq genomes not in the GTDB rs202 database used to benchmark orpheum accuracy.
{#tbl:refseq_acc tag="S3"}

## Practical considerations for building k~aa~-mer metapangenomes

### Open reading frame prediction with orpheum

The RAM used to run orpheum is dictated by the database size, as the database is loaded into to memory while it is running.
The GTDB rs202 nodegraph was 94 GB in size, and the RAM required to run orpheum never exceed 97GB, which makes database distribution and orpheum execution available on high performance compute clusters and other remote computers.
Alternatively, species level databases were ~5 Mb in size, reducing the RAM and CPU time needed to run orpheum.

We demonstrated that orpheum is better able to predict open reading frames in genomes that have species-level representatives in the GTDB database. 
To asses whether this criteria is satisfied by a query genome without performing genome assembly, we recommend sourmash gather [@doi:10.1101/2022.01.11.475838]. 
Sourmash gather will estimate the fraction of sequencing reads in a genome or metagenome that match to genomes in GTDB by comparing long nucleotide k-mers in the query against those in the database [@doi:10.1101/2022.01.11.475838].
Alternatively, the tool SingleM could be used to perform this task (https://github.com/wwood/singlem). 
SingleM estimates the taxonomic composition of sequencing reads by identifying fragments of single copy marker genes in short reads and comparing them against a database of taxonomically labelled sequences.

These strategies may also be useful to predetermine the set of species-level databases to use for ORF prediction.

### K~aa~-mer pangenome construction

One consideration is for k~aa~-mer sketch creation is the scaled value.
The scaled parameter controls the fraction on k~aa~-mers included in each sketch.
We have found that a scaled value of 100 works well for comparing proteomes [@url:https://bluegenes.github.io/2022-paper-protein-kmers].
For a subset of pangenomes benchmarked in results section, "K-mer methods accurately predict open reading frames in short sequencing reads", we tested scaled = 1 and scaled = 100 (n = 8; s__Akkermansia muciniphila, s__Chlamydia trachomatis, s__Faecalibacterium prausnitzii_D s__Gemmiger formicilis, s__Geothrix sp903857495, s__Methanobrevibacter_A-sp900314695, s__Thermus scotoductus, s__UBA2262 sp903869265).
We correlated the number of k~aa~-mers and genes per pangenome, the number of unique k~aa~-mers and genes per pangenome, and the similarity between genomes using k~aa~-mers or genes with sketches generated using a scaled of 1 or a scaled of 100.
We then performed a second order correlation to determine if both scaled values produced the same results.
All values were strongly significantly correlated with an R^2^ > 0.99 (p < 0.001), indicating that scaled 100 sketches captured the same patterns as scaled 1 sketches.
While we have not experimented with the upper bound of the scaled value that will still produce accurate results, using a scaled of 100 substantially decreased compute times.

