<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Taylor E. Reiter" />
  <meta name="author" content="N. Tessa Pierce-Ward" />
  <meta name="author" content="Luiz Irber" />
  <meta name="author" content="Olga Borisovna Botvinnik" />
  <meta name="author" content="C. Titus Brown" />
  <meta name="dcterms.date" content="2022-02-08" />
  <meta name="keywords" content="metapangenome, pangenome, assembly graph, reduced alphabet k-mer, open reading frame" />
  <title>Protein k-mers enable assembly-free microbial metapangenomics</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Protein k-mers enable assembly-free microbial metapangenomics" />
  <meta name="citation_title" content="Protein k-mers enable assembly-free microbial metapangenomics" />
  <meta property="og:title" content="Protein k-mers enable assembly-free microbial metapangenomics" />
  <meta property="twitter:title" content="Protein k-mers enable assembly-free microbial metapangenomics" />
  <meta name="dc.date" content="2022-02-08" />
  <meta name="citation_publication_date" content="2022-02-08" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Taylor E. Reiter" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0002-7388-421X" />
  <meta name="twitter:creator" content="@ReiterTaylor" />
  <meta name="citation_author" content="N. Tessa Pierce-Ward" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0002-2942-5331" />
  <meta name="twitter:creator" content="@saltyscientist" />
  <meta name="citation_author" content="Luiz Irber" />
  <meta name="citation_author_institution" content="Graduate Group in Computer Science, UC Davis" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0003-4371-9659" />
  <meta name="twitter:creator" content="@luizirber" />
  <meta name="citation_author" content="Olga Borisovna Botvinnik" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0003-4412-7970" />
  <meta name="twitter:creator" content="@olgabot" />
  <meta name="citation_author" content="C. Titus Brown" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0001-6001-2677" />
  <link rel="canonical" href="https://taylorreiter.github.io/2021-paper-metapangenomes/" />
  <meta property="og:url" content="https://taylorreiter.github.io/2021-paper-metapangenomes/" />
  <meta property="twitter:url" content="https://taylorreiter.github.io/2021-paper-metapangenomes/" />
  <meta name="citation_fulltext_html_url" content="https://taylorreiter.github.io/2021-paper-metapangenomes/" />
  <meta name="citation_pdf_url" content="https://taylorreiter.github.io/2021-paper-metapangenomes/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://taylorreiter.github.io/2021-paper-metapangenomes/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://taylorreiter.github.io/2021-paper-metapangenomes/v/d8372744f5957a17e452d7b065730e10aaa64703/" />
  <meta name="manubot_html_url_versioned" content="https://taylorreiter.github.io/2021-paper-metapangenomes/v/d8372744f5957a17e452d7b065730e10aaa64703/" />
  <meta name="manubot_pdf_url_versioned" content="https://taylorreiter.github.io/2021-paper-metapangenomes/v/d8372744f5957a17e452d7b065730e10aaa64703/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Protein k-mers enable assembly-free microbial metapangenomics</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://taylorreiter.github.io/2021-paper-metapangenomes/v/d8372744f5957a17e452d7b065730e10aaa64703/">permalink</a>)
was automatically generated
from <a href="https://github.com/taylorreiter/2021-paper-metapangenomes/tree/d8372744f5957a17e452d7b065730e10aaa64703">taylorreiter/2021-paper-metapangenomes@d837274</a>
on February 8, 2022.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Taylor E. Reiter</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-7388-421X">0000-0002-7388-421X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/taylorreiter">taylorreiter</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/ReiterTaylor">ReiterTaylor</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
· Funded by Grant XXXXXXXX
</small></p></li>
<li><p><strong>N. Tessa Pierce-Ward</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2942-5331">0000-0002-2942-5331</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/bluegenes">bluegenes</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/saltyscientist">saltyscientist</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
· Funded by NSF 1711984
</small></p></li>
<li><p><strong>Luiz Irber</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4371-9659">0000-0003-4371-9659</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/luizirber">luizirber</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/luizirber">luizirber</a><br>
<small>
Graduate Group in Computer Science, UC Davis; Department of Population Health and Reproduction, University of California, Davis
</small></p></li>
<li><p><strong>Olga Borisovna Botvinnik</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4412-7970">0000-0003-4412-7970</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/olgabot">olgabot</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/olgabot">olgabot</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>C. Titus Brown</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6001-2677">0000-0001-6001-2677</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/ctb">ctb</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<h1 id="introduction">Introduction</h1>
<p>Short read metagenomic sequencing has expanded our knowledge of microbial communities and diversity <span class="citation" data-cites="g228aulb hMM4KrP3 Lj5Ws5pH">[<a href="#ref-g228aulb" role="doc-biblioref">1</a>,<a href="#ref-hMM4KrP3" role="doc-biblioref">2</a>,<a href="#ref-Lj5Ws5pH" role="doc-biblioref">3</a>]</span>.
In particular, metagenome assembly and genome binning or annotation have produced catalogs of metagenome-assembled genomes and genes, revealing new species and functional potentially previously unobserved in cultured organisms <span class="citation" data-cites="F7RexqdF g228aulb hMM4KrP3">[<a href="#ref-g228aulb" role="doc-biblioref">1</a>,<a href="#ref-hMM4KrP3" role="doc-biblioref">2</a>,<a href="#ref-F7RexqdF" role="doc-biblioref">4</a>]</span>.</p>
<p>Along with advances in metagenome sequencing and analysis, the concept of metapangenomics has arisen as a framework for understanding how sets of metagenome-derived genes that are attributable to a group of organisms correlate with parameters in the environments in which they are sampled from <span class="citation" data-cites="tlupnHMS 1cNv4IK2 sv8w5xux">[<a href="#ref-tlupnHMS" role="doc-biblioref">5</a>,<a href="#ref-1cNv4IK2" role="doc-biblioref">6</a>,<a href="#ref-sv8w5xux" role="doc-biblioref">7</a>]</span>.
Metapangenomic methods borrow heavily from pangenome analysis.
Pangenomes comprise all genomic elements – usually open reading frames or genes – found within a group of organisms and reflect the metabolic and ecological plasticity of that group <span class="citation" data-cites="iRHOVXEg mrpq2yNB">[<a href="#ref-iRHOVXEg" role="doc-biblioref">8</a>,<a href="#ref-mrpq2yNB" role="doc-biblioref">9</a>]</span>.
The pangenome is divided into core and accessory genes, where core genes are shared by almost all members in the group and accessory genes are not.
Core genes often encode primary metabolism or other functions necessary for a group to live in a given environment <span class="citation" data-cites="TzhWL2dj">[<a href="#ref-TzhWL2dj" role="doc-biblioref">10</a>]</span>, while accessory genes encode functions that facilitate adaptation to changing environments <span class="citation" data-cites="mrpq2yNB">[<a href="#ref-mrpq2yNB" role="doc-biblioref">9</a>]</span>.
The size of the pangenome reflects the diversity of the organisms in a pangenome (population size, number of organisms sampled) as well as the ability of those organisms to adapt to different niches <span class="citation" data-cites="iRHOVXEg">[<a href="#ref-iRHOVXEg" role="doc-biblioref">8</a>]</span>.
Open pangenomes are those which increase indefinitely in size when adding new genomes, while closed pangenomes do not.</p>
<p>While pangenomes are traditionally inferred from isolate genomes, metapangenomics extends the ecological framework of pangenomics to metagenomes.
Metapangenomics gives insight into the genes that support specific environmental adaptations by applying pangenome methods to metagenome assembled genomes (MAGs) <span class="citation" data-cites="1cNv4IK2">[<a href="#ref-1cNv4IK2" role="doc-biblioref">6</a>]</span>, or by mapping metagenomes against isolate-inferred pangenomes <span class="citation" data-cites="tlupnHMS">[<a href="#ref-tlupnHMS" role="doc-biblioref">5</a>]</span>.
Both methods give valuable insight into the presence and distribution of functional content in natural microbial communities, but either may introduce biases associated with unknown sequencing content <span class="citation" data-cites="nSWbl26Q">[<a href="#ref-nSWbl26Q" role="doc-biblioref">11</a>]</span>.
MAGs are often incomplete or unrecoverable due to low sequencing coverage or large amounts of variation (SNPs, indels, rearrangements, horizontal gene transfer, sequencing error, etc.), both of which cause short read assemblers to produce unbinnable short contiguous sequences.
Unbinned sequences are disproportionately comprised of genomic islands and plasmids <span class="citation" data-cites="6Fmlbtl9">[<a href="#ref-6Fmlbtl9" role="doc-biblioref">12</a>]</span>, hot spots for evolution that support microbial adaptation to changing environments <span class="citation" data-cites="1HDOysM7N">[<a href="#ref-1HDOysM7N" role="doc-biblioref">13</a>]</span>.
In contrast, read mapping against isolate-inferred pangenomes may miss functional content present in the metagenome but missing from references, especially for species under represented or unrecorded in reference databases.</p>
<p>These issues are not exclusive to metapangenome inference, and many recently developed analysis strategies overcome some of these biases.
These techniques largely rely on k-mers, words of length <em>k</em> in DNA or protein sequences.
Metagenome k-mer profiles contain all sequences in a metagenome, including those which may not assemble or bin, or which aren’t in reference databases.
Long k-mers are also taxonomy-specific, where increasing k-mer length leads to sub-species discriminatory power <span class="citation" data-cites="mAjlUXL2">[<a href="#ref-mAjlUXL2" role="doc-biblioref">14</a>]</span> (CITE: TESSA).
These properties have popularized the use of k-mers for metagenome analysis, primarily through lightweight sketching and compact de Bruijn assembly graphs (cDBGs).
Lightweight sketching facilitates fast and accurate sequence comparisons between potentially large data sets through random but consistent sub-sampling <span class="citation" data-cites="mH9pzoIn IadOz4ol">[<a href="#ref-mH9pzoIn" role="doc-biblioref">15</a>,<a href="#ref-IadOz4ol" role="doc-biblioref">16</a>]</span>.
cDBGs maintain connectivity between k-mers and organize them into species-specific neighborhoods <span class="citation" data-cites="jroubkzG VsLophuT">[<a href="#ref-jroubkzG" role="doc-biblioref">17</a>,<a href="#ref-VsLophuT" role="doc-biblioref">18</a>]</span>.</p>
<!--
Metapangenomics is reliant on *de novo* metagenome analysis, but both assembly and binning introduce biases into analysis [@doi:10.1186/s13059-020-02066-4; @doi:10.1038/nmeth.4458; @doi:10.1101/2021.07.12.451567; @doi:10.1038/s41396-018-0081-5; @doi:10.1099/mgen.0.000436; @doi:10.1101/2021.05.04.442591].
Low coverage or large amounts of variation (SNPs, indels, rearrangements, horizontal gene transfer, sequencing error, etc.) cause assemblers to break contiguous sequences, producing short fragments or unassembled reads that are too short to be binned into MAGs (CITE).
-->
<p>To more fully represent the functional potential in metapangenomes, we present an analysis approach that relies on amino acid k-mers and assembly graph queries to estimate microbial (meta)pangenomes.
This approach for metapangenome estimation is minimally reliant on reference databases and is assembly-free.</p>
<h1 id="results">Results</h1>
<p>In an effort to reconstruct metapangenomes without loss of information from assembly and binning <span class="citation" data-cites="VsLophuT lsbnKJf8 A7OCZbgE jroubkzG 6Fmlbtl9 lQHniSOY">[<a href="#ref-6Fmlbtl9" role="doc-biblioref">12</a>,<a href="#ref-jroubkzG" role="doc-biblioref">17</a>,<a href="#ref-VsLophuT" role="doc-biblioref">18</a>,<a href="#ref-lsbnKJf8" role="doc-biblioref">19</a>,<a href="#ref-A7OCZbgE" role="doc-biblioref">20</a>,<a href="#ref-lQHniSOY" role="doc-biblioref">21</a>]</span>, we demonstrate a pipeline that relies on k-mers and assembly graphs for metapangenome estimation (<strong>Figure <a href="#fig:overview">1</a></strong>).
We first show that amino acid k-mers accurately estimate microbial pangenomes by comparing amino acid profiles of proteomes (translated coding domain sequences) against the proteomes themselves (<strong>Figure <a href="#fig:overview">1</a> A</strong>).
To derive amino acid k-mers directly from shotgun metagenome reads, we next demonstrate the accuracy of a tool called orpheum for open reading frame prediction from short sequencing reads (<strong>Figure <a href="#fig:overview">1</a> B</strong>).
We use assembly graph genome queries to retrieve species-specific reads from the metagenome, predict open reading frames from those reads using orpheum, and build a metapangenome using protein k-mers (<strong>Figure <a href="#fig:overview">1</a> C</strong>).
We then apply this method to species present over time in a time series metagenomes from a human gut microbiome.</p>
<div id="fig:overview" class="fignos">
<figure>
<img src="images/metapangenome_paper_overview.png" alt="Figure 1: Overview of the pipeline used to build metapangenomes. Approaches that were developed or tested in this manuscript are outlined in grey. A) We tested whether degenerate k-mer alphabets could accurately represent bacterial and archaeal pangenomes. Using genomes annotated with prokka, we compared pangenomes built with roary, a field-standard pipeline, against pangenomes built with degenerate k-mer alphabet sketches. B) We tested whether open reading frames could be predicted directly from short sequencing reads using the tool orpheum. This panel is modified from [22]. C) We combined this approaches with metagenome assembly graph genome queries to estimate metapangenomes directly from metagenomes without assembly or binning. The blue and orange lines correspond to steps tested in panels A and B." /><figcaption aria-hidden="true"><span>Figure 1:</span> <strong>Overview of the pipeline used to build metapangenomes.</strong>
Approaches that were developed or tested in this manuscript are outlined in grey.
<strong>A)</strong> We tested whether degenerate k-mer alphabets could accurately represent bacterial and archaeal pangenomes. Using genomes annotated with prokka, we compared pangenomes built with roary, a field-standard pipeline, against pangenomes built with degenerate k-mer alphabet sketches.
<strong>B)</strong> We tested whether open reading frames could be predicted directly from short sequencing reads using the tool orpheum. This panel is modified from <span class="citation" data-cites="TO2yVgRA">[<a href="#ref-TO2yVgRA" role="doc-biblioref">22</a>]</span>.
<strong>C)</strong> We combined this approaches with metagenome assembly graph genome queries to estimate metapangenomes directly from metagenomes without assembly or binning.
The blue and orange lines correspond to steps tested in panels <strong>A</strong> and <strong>B</strong>.</figcaption>
</figure>
</div>
<h2 id="reduced-alphabet-k-mers-accurately-estimate-characteristics-of-microbial-pangenomes">Reduced alphabet k-mers accurately estimate characteristics of microbial pangenomes</h2>
<p>Pangenomes from isolates are typically built by assembling each isolate genome and predicting genes (open reading frames), clustering gene sequences from all genomes into a non-redundant set, and estimating the presence/absence or abundance of each gene in each genome.
To determine whether bacterial and archaeal pangenomes could be constructed from reduced alphabet k-mers, we compared pangenomes estimated from genes against those estimated from k-mers (amino acid, dayhoff, and hydrophobic-polar).
We compared pangenomes from 23 species belonging to 23 phyla in the GTDB taxonomy <span class="citation" data-cites="vFApiGX2">[<a href="#ref-vFApiGX2" role="doc-biblioref">23</a>]</span>, with pangenome size ranging from 20-972 genomes (mean = 203 genomes, median = 44 genomes).
For each pangenome, we compared the total number of genes to the total number of k-mers, and the number of unique genes to the number of distinct k-mers within each genome.
We also tested the similarity of presence/absence profiles between pangenomes constructed with different methods using the Mantel test.</p>
<p>For these three metrics, performance varied minimally across encodings and k-mer sizes (<strong>Figure <a href="#fig:violin_fig">2</a></strong>).
This is likely because the genomes of the same species are closely related, so any reduced alphabet k-mer is sufficient to overcome minor genomic variations such as those introduced by codon degeneracy or evolutionary drift (CITE?).
Given that neither encoding nor k-mer size impacted these performance metrics, we selected protein k-mers with k = 10 to complete the rest of our analysis.
Protein k-mers of length 10 have recently been shown to perform well for comparisons across variable taxonomic distances (CITE: TESSA).</p>
<ul>
<li>Should I compare this against nucleotide k-mers at all? Bc I think that’s the underlying assumption here, nucleotides don’t work for this stuff. Tessa stuff sort of already shows this.</li>
</ul>
<div id="fig:violin_fig" class="fignos">
<figure>
<img src="images/violin_fig.png" alt="Figure 2: K-mer size and encoding do not impact pangenome estimation with k-mers. Violin plots representing the distribution of R2 values for linear models (Total, Unique) or statistic values for mantel tests (Mantel) calculated for each pangenome. Total corresponds to correlations between the total number of distinct genes and k-mers in a genome. Unique corresponds to correlations between the number of unique genes and k-mers in genome. Mantel corresponds to mantel tests between the gene and k-mer presence-absence matrices." /><figcaption aria-hidden="true"><span>Figure 2:</span> <strong>K-mer size and encoding do not impact pangenome estimation with k-mers.</strong> Violin plots representing the distribution of R<sup>2</sup> values for linear models (Total, Unique) or statistic values for mantel tests (Mantel) calculated for each pangenome. <em>Total</em> corresponds to correlations between the total number of distinct genes and k-mers in a genome. <em>Unique</em> corresponds to correlations between the number of unique genes and k-mers in genome. <em>Mantel</em> corresponds to mantel tests between the gene and k-mer presence-absence matrices.</figcaption>
</figure>
</div>
<p>While performance across metrics was similar for encodings and k-mer sizes, it varied dramatically for different pangenomes: both k-mers and genes are highly correlated for some pangenomes and are not correlated for others (<strong>Figure <a href="#fig:violin_fig">2</a></strong>).
We investigated pangenomes more closely to determine the source of the poor correlations and found that they were caused by the presence of many frameshifted proteins, one of many potential criteria for exclusion of GenBank genomes from RefSeq.
For example, <em>Leptospira interrogans</em> had an R<sup>2</sup> of 0.12 between the total number of genes and k-mers in genomes in the pangenome, but 21 of 317 genomes contained frameshifted proteins.
Removing these genomes increased the R<sup>2</sup> to 0.87 (<strong>Figure <a href="#fig:frameshift_fig">3</a> A</strong>).
This trend was consistent across pangenomes, where pangenomes with one or more frameshift-excluded genome had significantly lower R<sup>2</sup> values between total number of genes and k-mers per genome than pangenomes without (Welch Two Sample t-test, estimate = -0.36, p = 0.003) (<strong>Figure <a href="#fig:frameshift_fig">3</a> B</strong>).
Other RefSeq exclusion criteria did not impact the correlation between the total genes and k-mers per genome for a given pangenome.</p>
<div id="fig:frameshift_fig" class="fignos">
<figure>
<img src="images/frameshift_fig.png" alt="Figure 3: Genomes that are excluded from RefSeq for having many frameshifted proteins reduce similarity between gene- and k-mer-based pangenomes. A) Scatter plot of the total number of genes and k-mers per genome for the species Leptospira interrogans, where each point represents a single genome in the pangenome. Removing genomes flagged with RefSeq exclusion criteria “many frameshifted proteins” improves the correlation between these variables. B) Box plot of R2 values between the total number of genes and k-mers per genome. Pangenomes that contain genomes with the RefSeq exclusion criteria of “many frameshifted proteins” have significantly lower R2 values" /><figcaption aria-hidden="true"><span>Figure 3:</span> <strong>Genomes that are excluded from RefSeq for having many frameshifted proteins reduce similarity between gene- and k-mer-based pangenomes.</strong>
<strong>A)</strong> Scatter plot of the total number of genes and k-mers per genome for the species <em>Leptospira interrogans</em>, where each point represents a single genome in the pangenome.
Removing genomes flagged with RefSeq exclusion criteria “many frameshifted proteins” improves the correlation between these variables.
<strong>B)</strong> Box plot of R<sup>2</sup> values between the total number of genes and k-mers per genome.
Pangenomes that contain genomes with the RefSeq exclusion criteria of “many frameshifted proteins” have significantly lower R<sup>2</sup> values</figcaption>
</figure>
</div>
<p>We next investigated whether other pangenome metrics were well correlated between our k-mer-based and the gene-based method roary using pangenomes that did not contain genomes excluded from RefSeq for having many frameshifted proteins (see Methods for details).
For these 13 pangenomes, the percent of k-mers or genes predicted to be part of the core, shell, or cloud pangenome was strongly correlated (<strong>Figure <a href="#fig:pg_fig">4</a></strong>).
We also compared whether pangenomes would be designated as open or closed by calculating the alpha value for the Heaps law model <span class="citation" data-cites="MSOGoMuF">[<a href="#ref-MSOGoMuF" role="doc-biblioref">24</a>]</span>.
Alpha values were strongly correlated between gene- and k-mer based pangenomes (<strong>Figure <a href="#fig:pg_fig">4</a></strong>).</p>
<div id="fig:pg_fig" class="fignos">
<figure>
<img src="images/pg_fig.png" style="height:2.5in" alt="Figure 4: Pangenome metrics strongly correlate between gene- and k-mer-based pangenomes. Pangenome categories core, shell, and cloud refer to genes or k-mers shared between the majority (&gt;95%), some, or singleton genomes in the pangenome. Alpha is a value from Heaps law used to estimate whether a pangenome is open or closed." /><figcaption aria-hidden="true"><span>Figure 4:</span> <strong>Pangenome metrics strongly correlate between gene- and k-mer-based pangenomes.</strong> Pangenome categories core, shell, and cloud refer to genes or k-mers shared between the majority (&gt;95%), some, or singleton genomes in the pangenome. Alpha is a value from Heaps law used to estimate whether a pangenome is open or closed.</figcaption>
</figure>
</div>
<p>Taken together, these results show that reduced alphabet k-mers can accurately estimate key characteristics of pangenomes from bacterial and archaeal genomes.</p>
<h2 id="k-mer-methods-accurately-predicts-open-reading-frames-in-short-sequencing-reads">K-mer methods accurately predicts open reading frames in short sequencing reads</h2>
<p>We next sought to determine whether open reading frames could be accurately predicted directly from short sequencing reads, as this would enable k-mer-based pangenome analysis without assembly.
Without accurate open reading frame prediction, reads would need to be translated into all six translation frames prior to k-mer decomposition.
This would inflate the number of k-mers and decrease similarity between genomes.</p>
<p>We evaluated whether orpheum, a tool recently developed to predict open reading frames in Eukaryotic short reads <span class="citation" data-cites="TO2yVgRA">[<a href="#ref-TO2yVgRA" role="doc-biblioref">22</a>]</span>, could also perform this task in bacterial and archaeal sequences.
Orpheum predicts open reading frames by comparing reduced alphabet k-mers in six frame translations of short sequencing reads against those in a database (Jaccard containment) and assigns an open reading frame as coding if containment exceeds a user-defined threshold <span class="citation" data-cites="TO2yVgRA">[<a href="#ref-TO2yVgRA" role="doc-biblioref">22</a>]</span>.
To evaluate orpheum, we constructed a database containing all k-mers in coding domain sequences from genomes in GTDB rs202.
Using representative genomes from the 23 species above, as well as 20 additional RefSeq genomes not in the GTDB rs202 database, we simulated short sequencing reads either from coding domain sequences or non-coding sequences and used these reads to test orpheum.</p>
<p>Using default parameters, orpheum accurately separated coding from non-coding reads when reads were simulated from genomes in GTDB (<strong>Figure <a href="#fig:orpheum_fig">5</a> A</strong>).
On average, XX% of reads that were coding were predicted to be non-coding, while XX% of reads that were non-coding were predicted to be coding.
For reads simulated from genomes not in GTDB, orpheum recovered the majority of coding reads when genomes of the same species were in the database (<strong>Figure <a href="#fig:orpheum_fig">5</a> A,B</strong>).
On average, XX% of reads that were coding were predicted to be non-coding, while XX% of reads that were non-coding were predicted to be coding.
Accuracy decreased with increasing taxonomic distance between the query genome and the closest relative in the database (<strong>Figure <a href="#fig:orpheum_fig">5</a> B</strong>).</p>
<p>For genomes that had at least species-level representatives in GTDB, the largest source of error was non-coding reads being predicted as coding (<strong>Figure <a href="#fig:orpheum_fig">5</a> A</strong>).
We hypothesized that these reads originated from pseudogenes as these sequences would likely not be annotated as coding in the genomes from which the reads were simulated from, but may retain some k-mers contained in the database.
To assess this hypothesis, we used annotation files produced by the NCBI Prokaryotic Genome Annotation Pipeline (PGAP), which annotates pseudogenes, for the 23 genomes for which these files were available <span class="citation" data-cites="b3TnpDgh UUdrBbKz">[<a href="#ref-b3TnpDgh" role="doc-biblioref">25</a>,<a href="#ref-UUdrBbKz" role="doc-biblioref">26</a>]</span>.
On average, 12.4% (SD = 13.8%) of non-coding reads that were predicted to be coding fell within pseudogenes annotated by the PGAP pipeline.
We then BLASTed a subset of the remaining non-coding reads that were predicted to be coding against the NCBI nr database.
All reads we investigated had at least one match at 100% identity to protein sequences in the database, suggesting our test genomes contained additional pseudogenes not annotated by PGAP, or that the software we used to predict open reading frames missed some coding sequences (see Methods).
<!--Indeed, the worst performer was *Mycoplasmopsis bovis*, with 23% of non-coding reads predicted to be coding. 
*M. bovis* causes bovine tuberculosis and its genome contains signatures of genomic decay and pseudogenization [@doi:10.3389/fmicb.2017.02389], processes that may be linked to bacterial host specialization [@doi:10.1101/cshperspect.a010041; @doi:10.1093/gbe/evv135].-->
Because this method of open reading frame prediction cannot distinguish pseudogenes, it may not be appropriate for species with many pseudogenes.</p>
<p>Some coding sequences were also predicted to be non-coding.
We hypothesized that this was caused by sequencing error introduced into the simulated reads.
We mapped the simulated reads against the coding domain sequences from which they were derived and calculated mapping error rates.
While all reads mapped, the error rate was higher for reads that were predicted to be non-coding than those predicted to be coding (Welch Two Sample t-test, estimate = 0.00523, p &lt; 0.001).</p>
<p>Protein k-mers from predicted open reading frames in the simulated short sequencing reads recapitulated similarity between genomic coding domain sequences.
We estimated the Jaccard similarity between genomes using protein k-mers (<em>k</em> = 10) from annotated coding domain sequences, and compared this against Jaccard similarity between genomes using protein k-mers from predicted open read frames in the simulated short sequencing reads.
Genomes that were most similar in one matrix were also most similar in another matrix (Mantel statistic = 0.9975, p &lt; 0.001).
The average similarity among all pairwise comparisons for the coding domain sequences was 2.6%, and this decreased to 2.5% when using the open reading frames predicted from reads.
This demonstrates that information recovered from open reading frame prediction from short read is similar to that derived directly from the genome sequence.</p>
<p>The majority of predictive capability originated from species-level databases.
We performed ORF prediction using just species-level databases for genomes that had at least a species-level representative in GTDB, and compared this against ORF prediction using the full GTDB database.
On average, there was no change between the percent of reads derived from coding domain sequences when a species-level database was used versus when all of GTDB was used to predict open reading frames (<strong>Figure <a href="#fig:orpheum_db">9</a></strong>).</p>
<p>Decreasing the Jaccard containment threshold increased the sensitivity and specificity of ORF prediction when there are no closely related genomes in the database (<strong>Figure <a href="#fig:orpheum_fig">5</a> C, Table <a href="#tbl:threshold">1</a></strong>).
The Jaccard containment threshold controls the final prediction of coding vs. non-coding, as well as the the number of open reading frames which a read is translated into.
On average, increasing the rank of the closest taxonomic relative in the database by one taxonomic level decreased the optimal Jaccard containment threshold by 0.13.</p>
<div id="tbl:threshold" class="tablenos">
<table id="tbl:threshold" style="width:100%;">
<caption><span>Table 1:</span> Jaccard containment thresholds that maximize the Youden’s index depending on the taxonomic rank of the closest relative in GTDB.
</caption>
<colgroup>
<col style="width: 19%" />
<col style="width: 18%" />
<col style="width: 21%" />
<col style="width: 21%" />
<col style="width: 18%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Jaccard threshold</th>
<th style="text-align: center;">closest rank</th>
<th style="text-align: center;">mean sensitivity</th>
<th style="text-align: center;">mean specificity</th>
<th style="text-align: center;">mean Youden’s index</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">0.47</td>
<td style="text-align: center;">genome</td>
<td style="text-align: center;">0.988</td>
<td style="text-align: center;">0.971</td>
<td style="text-align: center;">0.959</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.39</td>
<td style="text-align: center;">species</td>
<td style="text-align: center;">0.941</td>
<td style="text-align: center;">0.961</td>
<td style="text-align: center;">0.902</td>
</tr>
<tr class="odd">
<td style="text-align: center;">0.17</td>
<td style="text-align: center;">genus</td>
<td style="text-align: center;">0.790</td>
<td style="text-align: center;">0.862</td>
<td style="text-align: center;">0.653</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.07</td>
<td style="text-align: center;">family</td>
<td style="text-align: center;">0.593</td>
<td style="text-align: center;">0.878</td>
<td style="text-align: center;">0.471</td>
</tr>
</tbody>
</table>
</div>
<p>Overall, these results show that open reading frames can be accurately determined from short sequencing reads when closely related proteomes are available.</p>
<div id="fig:orpheum_fig" class="fignos">
<figure>
<img src="images/orpheum_fig.png" alt="Figure 5: Orpheum correctly assigned short sequencing reads as coding or non-coding and selects the correct open reading frame. A) Percent of simulated coding or non-coding sequences predicted as coding, non-coding, or discarded based on quality metrics (see methods). Genomes are split by those in GTDB and those not in GTDB. Genomes not in GTDB are labelled by taxonomic assignment from GTDB-tk. Predictions were made using default parameters (Jaccard containment = 0.5). B) Boxplots of the percent of coding reads that were recovered by Orpheum, separated by the level of taxonomic assignment achieved by GTDB-Tk. Orpheum recovers more coding sequences when there are closely related genomes in the database. C) Receiver operating curves for the Jaccard containment thresholds. Curves are separated by the level of taxonomic assignment achieved by GTDB-Tk, and values are averaged across all genomes that fell within those categories. The best Jaccard threshold decreases when there are fewer closely related genomes in the database. D) Databases constructed of only closely-related genomes recover the majority of coding sequences, but including increasingly distantly related genomes improves total coding recall." /><figcaption aria-hidden="true"><span>Figure 5:</span> <strong>Orpheum correctly assigned short sequencing reads as coding or non-coding and selects the correct open reading frame.</strong>
<strong>A)</strong> Percent of simulated coding or non-coding sequences predicted as coding, non-coding, or discarded based on quality metrics (see methods).
Genomes are split by those in GTDB and those not in GTDB.
Genomes not in GTDB are labelled by taxonomic assignment from GTDB-tk.
Predictions were made using default parameters (Jaccard containment = 0.5).
<strong>B)</strong> Boxplots of the percent of coding reads that were recovered by Orpheum, separated by the level of taxonomic assignment achieved by GTDB-Tk.
Orpheum recovers more coding sequences when there are closely related genomes in the database.
<strong>C)</strong> Receiver operating curves for the Jaccard containment thresholds.
Curves are separated by the level of taxonomic assignment achieved by GTDB-Tk, and values are averaged across all genomes that fell within those categories.
The best Jaccard threshold decreases when there are fewer closely related genomes in the database.
<strong>D)</strong> Databases constructed of only closely-related genomes recover the majority of coding sequences, but including increasingly distantly related genomes improves total coding recall.</figcaption>
</figure>
</div>
<ul>
<li>Should/do I have to compare these results against FragGeneScan?</li>
</ul>
<h2 id="k-mer-based-metapangenomics-combined-with-assembly-graphs">K-mer-based metapangenomics combined with assembly graphs …</h2>
<p>Given that amino acid k-mers accurately estimated pangenomes, and that the correct open reading frame could be predicted reliably from short sequencing data, we next combined these approaches to perform metapangenome analysis from short read shotgun metagenomes.
We used 12 metagenomes from a single individual sampled over the course of a year by the Integrated Human Microbiome Project (iHMP) <span class="citation" data-cites="UotiKLjf">[<a href="#ref-UotiKLjf" role="doc-biblioref">27</a>]</span>.
The individual was diagnosed with Crohn’s disease, a sub type of inflammatory bowel disease characterized by inflammation along the gastrointestinal tract.
The individual received three courses of antibiotics over the year and each course was separated by weeks without antibiotics (<strong>Figure <a href="#fig:metag_species">6</a></strong>).</p>
<div id="fig:metag_species" class="fignos">
<figure>
<img src="images/common_species_breakdown.png" style="height:3in" alt="Figure 6: Antibiotic courses and corresponding gut microbiome profiles for a single individual with Crohn’s disease. Fractional abundances are colored by species, with only the six species that accounted for greater than 2% of all metagenome reads displayed." /><figcaption aria-hidden="true"><span>Figure 6:</span> <strong>Antibiotic courses and corresponding gut microbiome profiles for a single individual with Crohn’s disease.</strong>
Fractional abundances are colored by species, with only the six species that accounted for greater than 2% of all metagenome reads displayed.</figcaption>
</figure>
</div>
<p>We estimated the metapangenome for each species that was detected in all 12 metagenomes and that accounted for at least 2% of reads across metagenomes, for a total of six metapangenomes (<strong>Figure <a href="#fig:metag_species">6</a></strong>).
To obtain all sequencing reads that originated from genomes of these species, we performed assembly graph genome queries <span class="citation" data-cites="VsLophuT">[<a href="#ref-VsLophuT" role="doc-biblioref">18</a>]</span>.
Assembly graphs contain all sequences in a metagenome, and assembly graph queries return sequences in the metagenome that are either in the query or nearby to the query in the graph.
Assembly graph genome queries return sequencing reads that originate from genomes in the metagenome that have as little as 0.1 Jaccard similarity (approximately 93% average nucleotide identity (ANI) (CITE: TESSA)) to the query genome <span class="citation" data-cites="VsLophuT">[<a href="#ref-VsLophuT" role="doc-biblioref">18</a>]</span>.
After retrieving reads in this way, we predicted open reading frames using orpheum.
We used species-level databases as these were successful in the context of isolate genomes not in the database (see above) and because they would be more likely to filter out reads beyond the species boundary (95% ANI <span class="citation" data-cites="12fAf7MKk">[<a href="#ref-12fAf7MKk" role="doc-biblioref">28</a>]</span>) that were returned by assembly graph queries.
Using the predicted amino acid sequences, we built metapangenomes for each of the six species (<strong>Figure <a href="#fig:metap_fig">7</a></strong>).</p>
<p>Unlike isolate genomes, metagenomes may contain a fraction of an organism’s genome if the metagenome was not sequenced deeply or if an organism was rare.
To calculate the core, shell, and cloud fractions and to estimate the openness of the metapangenome, we removed samples with fewer than 10,000 k-mers (<strong>Table <a href="#tbl:metap">2</a></strong>).</p>
<div id="tbl:metap" class="tablenos">
<table id="tbl:metap">
<caption><span>Table 2:</span> Metapangenome estimates for each species. <em>n</em> designates the number of metagenomes used to estimate the total, core, shell, cloud, and alpha values.
</caption>
<thead>
<tr class="header">
<th style="text-align: center;">species</th>
<th style="text-align: center;">n</th>
<th style="text-align: center;">total</th>
<th style="text-align: center;">core</th>
<th style="text-align: center;">shell</th>
<th style="text-align: center;">cloud</th>
<th style="text-align: center;">alpha</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><em>Bacteroides fragilis</em></td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">24819</td>
<td style="text-align: center;">56.3%</td>
<td style="text-align: center;">11.3%</td>
<td style="text-align: center;">32.4%</td>
<td style="text-align: center;">0.76</td>
</tr>
<tr class="even">
<td style="text-align: center;"><em>Bacteroides uniformis</em></td>
<td style="text-align: center;">9</td>
<td style="text-align: center;">32197</td>
<td style="text-align: center;">38.0%</td>
<td style="text-align: center;">22.3%</td>
<td style="text-align: center;">39.7%</td>
<td style="text-align: center;">0.73</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><em>Enterocloster bolteae</em></td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">23620</td>
<td style="text-align: center;">55.8%</td>
<td style="text-align: center;">18.3%</td>
<td style="text-align: center;">25.9%</td>
<td style="text-align: center;">0.66</td>
</tr>
<tr class="even">
<td style="text-align: center;"><em>Parabacteroides distasonis</em></td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">25789</td>
<td style="text-align: center;">42.4%</td>
<td style="text-align: center;">30.9%</td>
<td style="text-align: center;">26.8%</td>
<td style="text-align: center;">0.74</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><em>Parabacteroides merdae</em></td>
<td style="text-align: center;">6</td>
<td style="text-align: center;">19985</td>
<td style="text-align: center;">63.2%</td>
<td style="text-align: center;">9.6%</td>
<td style="text-align: center;">27.1%</td>
<td style="text-align: center;">0.82</td>
</tr>
<tr class="even">
<td style="text-align: center;"><em>Phocaeicola vulgatus</em></td>
<td style="text-align: center;">11</td>
<td style="text-align: center;">41005</td>
<td style="text-align: center;">30.3%</td>
<td style="text-align: center;">20.4%</td>
<td style="text-align: center;">49.2%</td>
<td style="text-align: center;">0.65</td>
</tr>
</tbody>
</table>
</div>
<p>Using our metapangenome approach, we identified interesting patterns in accessory gene presence associated with antibiotic exposure (<strong>Figure <a href="#fig:metap_fig">7</a></strong>).
For example, the <em>Phocaeicola vulgatus</em> metapangenome is stable for the first 11 weeks of sampling even during ciprofloxacin exposure,
but a portion of the accessory genome corresponding to XX% of the total metapangenome disappears at week 13, coinciding with metronidazole administration.
While a portion of the <em>Parabacteroides merdae</em> metapangenome is present in early samples, the full genome is only detected after metronidazole administration when the fractional abundance of <em>P. merdae</em> increases starting at week 13.
However, additional accessory elements are detected beginning at week 19, which coincides with a bloom of <em>P. merdae</em>.</p>
<p>In two bacterial species, antibiotic administration appears to spur on strain switching.
In <em>Bacteroides uniformis</em>, one set of accessory elements present from weeks 0 - 11 is replaced by a new set in weeks 25-36.
Similarly, in <em>Parabacteroides distasonis</em>, accessory elements present in weeks 4, 9, and 11 are replaced by new accessory elements in weeks 25-36.
Both switches occur during metronidazole administration after the bloom of <em>P. merdae</em> and <em>P. vulgatus</em>.</p>
<div id="fig:metap_fig" class="fignos">
<figure>
<img src="images/metap.png" alt="Figure 7: " /><figcaption aria-hidden="true"><span>Figure 7:</span> </figcaption>
</figure>
</div>
<ul>
<li>Do I need to compare these results against typical metapangenomics? like do de novo assembly, binning, prokka? etc?</li>
<li>What else, if anything, belongs in this section?</li>
</ul>
<h1 id="discussion">Discussion</h1>
<p>We present a method to perform assembly-free metapangenomics that is minimally reliant on reference databases.
We show that pangenome metrics like core, cloud, and shell pangenome fractions can be accurately estimated with long amino acid k-mers.
We then demonstrate accurate prediction of open reading frames in highly accurate short sequencing reads by comparing amino acid k-mers in all translation frames against a database of k-mers from all known bacterial and archaeal genomes in GTDB (rs202).
Combining these tools enables pangenome estimation directly from quality controlled short sequencing reads.
In the context of metagenomes, these approaches enable metapangenome estimation without the need to <em>de novo</em> assemble and bin sequences, eliminating common sources of lost sequencing variation (cite spacegraphcats). These techniques also reduce the dependence of metapangenomics on complete or comprehensive reference databases, which can be important for understudied environments.</p>
<p>The combination of these approaches is potentially most useful in the context of analyzing metagenome assembly graphs.
Assembly graphs like compact de Bruijn graphs (cDBG) capture all sequences in a metagenome, including sequences with high strain variation or low coverage, which may not be captured by other analysis methods.
A targeted query of an assembly graph, for example with a metagenome-assembled genome bin, can recover all sequencing reads in a metagenome that originate from all genomes of the same species (cite spacegraphcats).
While recovering these reads and assigning their taxonomic identity through graph queries is useful, many of the recovered reads cannot be assembled due to prolific sequencing variation attributable to strain diversity in the original microbial community.
Yet, the sequences represented by these un-assembleable reads often encode functional potential, some of which may be key to a microorganisms functioning within its ecosystem (cite sumner paper?; metachercant).
The approaches presented in this paper enable these sequences to be represented in metapangenome estimation.</p>
<ul>
<li>from #2 titus: it might be good (somewhere) to talk about how working with reads is better than working with cDBGs, because in regions of high error / high variation, the cDBG nodes or often shorter than reads.</li>
</ul>
<p>Long read sequencing of microbial communities stands to improve many of these challenges, particularly as lineage-resolved methods become mainstream (cite bickhart et al.).
Even as long read technologies improve, short read sequences continue to better capture strain diversity from a community (Cite Maureen?).
Even with long read references from the same community, many of these short reads do not map and do not assemble (cite Maureen).
The approaches presented here will allow these sequences to be included in pangenome estimation.</p>
<p>Practically, open reading frame prediction with orpheum can be executed on microbial illumina short read data sets.
<!-- Generalize? Include citation to olga paper here? -->
The RAM used to run orpheum is dictated by the database size, as the database is loaded into to memory while its running.
The GTDB rs202 nodegraph was 94 GB, and the RAM required to run orpheum never exceed 97GB, which makes database distribution and orpheum execution available on high performance compute clusters and other remote computers.
To reduce ram, this data structure could be improved XXX.
Alternatively, species level databases were ~5 Mb in size, reducing the RAM and CPU tiem needed to run orpheum.
<!-- Titus/Tessa? --></p>
<p>We demonstrated that orpheum is better able to predict open reading frames in genomes that have species-level representatives in the GTDB database.
To asses whether this criteria is satisfied by a query genome without performing genome assembly, we recommend sourmash gather. Sourmash gather will estimate the fraction of sequencing reads in a genome or metagenome that match to genomes in GTDB by comparing long nucleotide k-mers in the query against those in the database (cite gather paper).
Alternatively, the tool SingleM could be used to perform this task.
SingleM estimates the taxonomic composition of sequencing reads by identifying fragments of single copy marker genes in short reads and comparing them against a database of taxonomically labelled sequences.</p>
<p>These strategies may also be useful to predetermine the set of species-level databases to use for ORF prediction.
<!-- Interestingly, while shared k-mer content between distantly related genomes improves ORF prediction, using reduced alphabets did not improve ORF recall. -->
<!-- We only did this for panmers, not orpheum...so for. Could show e.g. within phlya or some small db.  --></p>
<p>Comparison between euks? Need to read orpheum paper.</p>
<p>PANMER discussion</p>
<ul>
<li>sourmash signature generation is rapid.</li>
<li>Exact matching scales (linearly?). May enable running on very large collections of genomes.</li>
<li>Exact matching of k-mers enables additions of new species without having to rerun everything.</li>
<li>Exact matching also allows direct comparisons to distantly related organisms. Unified framework for genome comparisons even when organisms are distantly related.</li>
<li>scaled is handy parameter to potentially enable even larger comparisons</li>
<li>sacrifice function – annotating k-mers with function is good future work.</li>
</ul>
<h3 id="other-points">Other points</h3>
<ul>
<li>While the number of genes per genome is increased for genomes with this exclusion criteria, there is no commensurate increase in the number of k-mers observed.
This suggests that the number of k-mers in a genome could be used to predict the expected range of predicted genes in a genome, and could be potentially used a quality control metric for annotated genomes.</li>
<li>While developed for the metapangenomics space, this study demonstrates that k-mer-based pangenomes will also work in isolate genomes. Given that building k-mer sketches and exact matching of k-mers between genomes is fast, this provides an alternative approach for building pangenomes.</li>
<li>De novo metagenome analysis probably dramatically improves ORF prediction because of the inclusion of these genomes in GTDB.</li>
<li>annotation is substantial drawback. Integrate potential of assembly graph annotation.</li>
</ul>
<h1 id="methods">Methods</h1>
<!-- Can I pretty plz keep these separate or do they need to be combined? -->
<p>All code is available at github.com/taylorreiter/2021-panmers (results section 1), github.com/taylorreiter/2021-orpheum-sim (results section 2), and https://github.com/taylorreiter/2021-metapangenome-example (results section 3).</p>
<h3 id="selection-of-benchmarking-species-for-pangenome-analysis">Selection of benchmarking species for pangenome analysis</h3>
<p>We selected a species representative for each of the 23 phyla in GTDB rs202 <span class="citation" data-cites="vFApiGX2">[<a href="#ref-vFApiGX2" role="doc-biblioref">23</a>]</span>.
To select representative species, we first filtered species with fewer than 20 representatives and greater than 1000 representatives.
While this approach scales beyond 1000 genomes, we elected to benchmark smaller sets to iterate over the potential parameter space more quickly.
Of species remaining after filtering, we selected the species within each phyla that had the largest number of genomes.
We downloaded these genomes from GenBank.
Species names are recorded in <strong>Figure <a href="#fig:tree_fig">8</a></strong>.</p>
<h3 id="calculating-the-gene-based-pangenome-with-roary">Calculating the gene-based pangenome with roary</h3>
<p>To calculate the gene-based pangenome, we first annotated each genome using prokka <span class="citation" data-cites="JrcoonJ8">[<a href="#ref-JrcoonJ8" role="doc-biblioref">29</a>]</span>.
We then used the resulting GFF annotations files to calculate the pangenome with roary using default settings <span class="citation" data-cites="1cFZOigd">[<a href="#ref-1cFZOigd" role="doc-biblioref">30</a>]</span>.</p>
<h3 id="calculating-the-k-mer-based-pangenome-with-sourmash">Calculating the k-mer based pangenome with sourmash</h3>
<p>To calculate k-mer based pangenomes, we used sourmash <code>sketch</code> to generate signatures from the bakta-predicted amino acid sequences (<code>.faa</code> files) <span class="citation" data-cites="ExvjTLL3">[<a href="#ref-ExvjTLL3" role="doc-biblioref">31</a>]</span>.
We used the protein alphabet (k = 7, 8, 9, 10, 11), dayhoff alphabet (k = 13, 15, 17), and the hydrophobic-polar alphabet (k = 27, 31).
All signatures were calculated with a scaled value of 100.
The scaled parameter controls the fraction of the total k-mers represented by the sketch; a scaled value of 100 indicates that 1/100th of the distinct k-mers in a genome were included in each sketch.
We converted signatures from json format into a genome x hash presence-absence matrix.</p>
<h3 id="correlating-gene-based-and-k-mer-based-pangenomes">Correlating gene-based and k-mer based pangenomes</h3>
<p>Using the presence-absence matrices for the gene-based and k-mer-based pangenomes, we correlated total genes/k-mers observed per genome and total unique genes/k-mers observed per genome for each species.
We used the <code>rowSums()</code> function in R to determine the number of genes/unique genes per matrix, then used the <code>lm()</code> function with default parameters to correlate the values.
We also used the Mantel test to determine whether genomes that were most similar in the gene presence-absence matrix were also most similar in the k-mer presence-absence matrix.
We used the <code>mantel()</code> function in the R vegan package to perform this test <span class="citation" data-cites="6ZxBxq9m">[<a href="#ref-6ZxBxq9m" role="doc-biblioref">32</a>]</span>.
We used distance matrices calculated with the <code>dist()</code> function using the parameter <code>method = "binary"</code> as input to the mantel test.</p>
<h3 id="generating-standard-pangenome-metrics-with-pagoo">Generating standard pangenome metrics with pagoo</h3>
<p>The pagoo R package provides functions to analyze bacterial pangenomes <span class="citation" data-cites="10u9VaxL4">[<a href="#ref-10u9VaxL4" role="doc-biblioref">33</a>]</span>.
We used this package to generate standard pangenome metrics and visualizations.
These metrics are based on the presence-absence matrices generated above and include calculation of the core, shell, and cloud genome sizes and estimation of the alpha value in Heaps law for estimation of pangenome openness.</p>
<h3 id="augmenting-benchmarking-species-set-to-include-genomes-not-in-gtdb-for-open-reading-frame-prediction">Augmenting benchmarking species set to include genomes not in GTDB for open reading frame prediction</h3>
<p>We next generated a benchmarking data set for open reading frame prediction.
We selected a genome from each of the 23 species evaluated above, choosing the GTDB rs202 representative genome for each species.
Genome accessions are recorded in <strong>Table <a href="#tbl:gtdb_acc">4</a></strong>.
Given that open reading frame prediction relies on a database, and we used k-mers in GTDB rs202 to generate this database, we also wanted to select genomes that were not in GTDB to evaluate this method.
We determined the bacterial and archaeal genomes that were added to RefSeq after the construction of GTDB rs202 (April 2021-November 2021).
From this set, we selected a representative genome from each of the distinct NCBI phyla represented among these genomes, 20 in total.
Genome accessions are recorded in <strong>Table <a href="#tbl:refseq_acc">5</a></strong>.
We then ran GTDB-tk on these genomes to predict the GTDB taxonomy of each.</p>
<h3 id="simulating-coding-domain-sequence-and-non-coding-domain-sequence-reads-with-polyester">Simulating coding domain sequence and non coding domain sequence reads with polyester</h3>
<p>We next created a labelled data set of simulated reads that were generated from either coding domain sequences (CDS) or non-coding regions within each genome.
We annotated the genomes with bakta to produce CDS ranges <span class="citation" data-cites="wS8KrahX">[<a href="#ref-wS8KrahX" role="doc-biblioref">34</a>]</span>, and used polyester to simulate reads from CDS or non-coding regions <span class="citation" data-cites="LVpd95E4">[<a href="#ref-LVpd95E4" role="doc-biblioref">35</a>]</span>.
We used the default short read error profile within polyester.</p>
<h3 id="determining-short-read-open-reading-frames-with-orpheum">Determining short read open reading frames with orpheum</h3>
<p>We used the orpheum tool to predict open reading frames from simulated short reads <span class="citation" data-cites="TO2yVgRA">[<a href="#ref-TO2yVgRA" role="doc-biblioref">22</a>]</span>.
Orpheum was developed to predict open reading frames in short RNA-seq reads from Eukaryotic organisms without a reference genome or transcriptome sequence <span class="citation" data-cites="TO2yVgRA">[<a href="#ref-TO2yVgRA" role="doc-biblioref">22</a>]</span>.
Orpheum perform six-frame translation on nucleotide sequencing reads, calculates k-mers in an amino acid, dayhoff, or hydrophobic-polar encoding at the designated k-mer length, and then estimates the Jaccard similarity between k-mers in each translation frame and a database.
It then selects all open reading frames based on a Jaccard similarity threshold, and returns those reads as translated amino acid sequences.
Open reading frames are excluded if they contain stop codons, low complexity sequences, or if the read is too short to perform translation.
Reads are designated as non-coding if they don’t reach the Jaccard similarity threshold and are not excluded for other reasons.
<!-- Is it jaccard similarity? It's whichever one makes sense and I always get them confused --></p>
<p>We constructed a database from GTDB rs202 using sourmash XXX and using a k-mer size of 10.
+ <span class="citation" data-cites="Tessa">[<a href="#ref-Tessa" role="doc-biblioref"><strong>Tessa?</strong></a>]</span> any relevant details would be very helpful :)</p>
<h3 id="metapangenome-analysis-of-ihmp-metagenomes">Metapangenome analysis of iHMP metagenomes</h3>
<p>We used sourmash, spacegraphcats, and orpheum to peform metapangenome analysis of 12 iHMP time series gut microbiomes captured by short read shotgun metagenomes <span class="citation" data-cites="19Z0U0zL4">[<a href="#ref-19Z0U0zL4" role="doc-biblioref">36</a>]</span>.
We downloaded samples HSM6XRQB, HSM6XRQI, HSM6XRQK, HSM6XRQM, HSM6XRQO, HSM67VF9, HSM67VFD, HSM67VFJ, HSM7CYY7, HSM7CYYD, HSM7CYY9, HSM7CYYB from ibdmdb.org.
We adapter and quality trimmed each sample with fastp (parameters <code>--detect_adapter_for_pe</code>, <code>--qualified_quality_phred 4</code>, <code>--length_required 31</code>, and <code>--correction</code>), removed human host sequencing reads with bbduk (parameters <code>k=31</code>, reference file https://drive.google.com/file/d/0B3llHR93L14wd0pSSnFULUlhcUk/edit?usp=sharing), and k-mer trimmed reads using khmer <code>trim-low-abund.py</code> (parameters <code>-C 3</code>, <code>-Z 18</code>, <code>-V</code>) <span class="citation" data-cites="G2CzfHHF">[<a href="#ref-G2CzfHHF" role="doc-biblioref">37</a>]</span>.
We then used sourmash gather to infer the taxonomic profile of each sample, using the GTDB rs202 database (<code>k</code> = 31, https://osf.io/w4bcm/) <span class="citation" data-cites="19Z0U0zL4">[<a href="#ref-19Z0U0zL4" role="doc-biblioref">36</a>]</span>.
We summarized the results to species-level using the GTDB taxonomy.
We retained species with a cumulative sum of at least 2% (sum of <code>f_unique_to_query</code>) across metagenome reads as query genomes.
We downloaded each genome from GenBank (<strong>Table <a href="#tbl:acc">3</a></strong>) and performed spacegraphcats assembly graph queries with each (parameters <code>ksize: 31</code>, <code>radius: 1</code>, <code>paired_reads: true</code>) <span class="citation" data-cites="VsLophuT">[<a href="#ref-VsLophuT" role="doc-biblioref">18</a>]</span>.
Using the returned reads, we predicted open reading frames using orpheum <code>translate</code> (parameters <code>--jaccard-threshold 0.39</code>, <code>--alphabet protein</code>, <code>--peptide-ksize 10</code>) and using species-level GTDB databases.
We sketched each set of translated reads using sourmash <code>sketch</code> (parameters <code>protein</code>, <code>-p k=10,scaled=100,protein</code>) <span class="citation" data-cites="ExvjTLL3">[<a href="#ref-ExvjTLL3" role="doc-biblioref">31</a>]</span>, converted each sketch to a csv file, and then combined csv files for a single query species across all metagenomes.
This long format csv was used as input for the R pangenome package pagoo, using the <code>pagoo()</code> function <span class="citation" data-cites="10u9VaxL4">[<a href="#ref-10u9VaxL4" role="doc-biblioref">33</a>]</span>.
We used pagoo methods <code>pg$gg_binmap()</code>, <code>pg$summary_stats()</code>, and <code>pg$pg_power_law_fit()</code> to visualize the pangenome, calculate the size of the core, shell, and cloud, and estimate alpha.</p>
<div id="tbl:acc" class="tablenos">
<table id="tbl:acc">
<caption><span>Table 3:</span> Query genome GTDB species names and GenBank accessions.
</caption>
<thead>
<tr class="header">
<th style="text-align: center;">species</th>
<th style="text-align: center;">accession</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><em>Parabacteroides distasonis</em></td>
<td style="text-align: center;">GCA_000162535.1</td>
</tr>
<tr class="even">
<td style="text-align: center;"><em>Enterocloster bolteae</em></td>
<td style="text-align: center;">GCF_003433765.1</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><em>Bacteroides fragilis</em></td>
<td style="text-align: center;">GCF_003458955.1</td>
</tr>
<tr class="even">
<td style="text-align: center;"><em>Parabacteroides merdae</em></td>
<td style="text-align: center;">GCF_003475305.1</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><em>Bacteroides uniformis</em></td>
<td style="text-align: center;">GCF_009020325.1</td>
</tr>
<tr class="even">
<td style="text-align: center;"><em>Phocaeicola vulgatus</em></td>
<td style="text-align: center;">GCF_009025805.1</td>
</tr>
</tbody>
</table>
</div>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-g228aulb" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline"><strong>A new view of the tree of life</strong> <div class="csl-block">Laura A Hug, Brett J Baker, Karthik Anantharaman, Christopher T Brown, Alexander J Probst, Cindy J Castelle, Cristina N Butterfield, Alex W Hernsdorf, Yuki Amano, Kotaro Ise, … Jillian F Banfield</div> <em>Nature Microbiology</em> (2016-05) <a href="https://doi.org/bpkh">https://doi.org/bpkh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nmicrobiol.2016.48">10.1038/nmicrobiol.2016.48</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27572647">27572647</a></div></div>
</div>
<div id="ref-hMM4KrP3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline"><strong>A genomic catalog of Earth’s microbiomes</strong> <div class="csl-block">Stephen Nayfach, Simon Roux, Rekha Seshadri, Daniel Udwary, Neha Varghese, Frederik Schulz, Dongying Wu, David Paez-Espino, I-Min Chen, Marcel Huntemann, … Emiley A Eloe-Fadrosh</div> <em>Nature Biotechnology</em> (2021-04) <a href="https://doi.org/ghjh4b">https://doi.org/ghjh4b</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41587-020-0718-6">10.1038/s41587-020-0718-6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33169036">33169036</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8041624">PMC8041624</a></div></div>
</div>
<div id="ref-Lj5Ws5pH" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline"><strong>Phage diversity, genomics and phylogeny</strong> <div class="csl-block">Moïra B Dion, Frank Oechslin, Sylvain Moineau</div> <em>Nature Reviews Microbiology</em> (2020-03) <a href="https://doi.org/ggkq9f">https://doi.org/ggkq9f</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41579-019-0311-5">10.1038/s41579-019-0311-5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32015529">32015529</a></div></div>
</div>
<div id="ref-F7RexqdF" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline"><strong>Community structure and metabolism through reconstruction of microbial genomes from the environment</strong> <div class="csl-block">Gene W Tyson, Jarrod Chapman, Philip Hugenholtz, Eric E Allen, Rachna J Ram, Paul M Richardson, Victor V Solovyev, Edward M Rubin, Daniel S Rokhsar, Jillian F Banfield</div> <em>Nature</em> (2004-03) <a href="https://doi.org/b85j5j">https://doi.org/b85j5j</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nature02340">10.1038/nature02340</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/14961025">14961025</a></div></div>
</div>
<div id="ref-tlupnHMS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">5. </div><div class="csl-right-inline"><strong>Linking pangenomes and metagenomes: the &lt;i&gt;Prochlorococcus&lt;/i&gt; metapangenome</strong> <div class="csl-block">Tom O Delmont, AMurat Eren</div> <em>PeerJ</em> (2018-01-25) <a href="https://doi.org/gczf4x">https://doi.org/gczf4x</a> <div class="csl-block">DOI: <a href="https://doi.org/10.7717/peerj.4320">10.7717/peerj.4320</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29423345">29423345</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5804319">PMC5804319</a></div></div>
</div>
<div id="ref-1cNv4IK2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">6. </div><div class="csl-right-inline"><strong>Global ecotypes in the ubiquitous marine clade SAR86</strong> <div class="csl-block">Adrienne Hoarfrost, Stephen Nayfach, Joshua Ladau, Shibu Yooseph, Carol Arnosti, Chris L Dupont, Katherine S Pollard</div> <em>The ISME Journal</em> (2020-01) <a href="https://doi.org/gns4sb">https://doi.org/gns4sb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41396-019-0516-7">10.1038/s41396-019-0516-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31611653">31611653</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6908720">PMC6908720</a></div></div>
</div>
<div id="ref-sv8w5xux" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">7. </div><div class="csl-right-inline"><strong>Meta-Pangenome: At the Crossroad of Pangenomics and Metagenomics</strong> <div class="csl-block">Bing Ma, Michael France, Jacques Ravel</div> <em>The Pangenome</em> (2020) <a href="https://doi.org/gns4r8">https://doi.org/gns4r8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/978-3-030-38281-0_9">10.1007/978-3-030-38281-0_9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32633911">32633911</a> · ISBN: 9783030382803</div></div>
</div>
<div id="ref-iRHOVXEg" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">8. </div><div class="csl-right-inline"><strong>Genome analysis of multiple pathogenic isolates of Streptococcus agalactiae: Implications for the microbial "pan-genome"</strong> <div class="csl-block">H Tettelin, V Masignani, MJ Cieslewicz, C Donati, D Medini, NL Ward, SV Angiuoli, J Crabtree, AL Jones, AS Durkin, … CM Fraser</div> <em>Proceedings of the National Academy of Sciences</em> (2005-09-27) <a href="https://doi.org/b9b426">https://doi.org/b9b426</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1073/pnas.0506758102">10.1073/pnas.0506758102</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/16172379">16172379</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1216834">PMC1216834</a></div></div>
</div>
<div id="ref-mrpq2yNB" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">9. </div><div class="csl-right-inline"><strong>Toward quantifying the adaptive role of bacterial pangenomes during environmental perturbations</strong> <div class="csl-block">Roth E Conrad, Tomeu Viver, Juan F Gago, Janet K Hatt, Stephanus N Venter, Ramon Rossello-Mora, Konstantinos T Konstantinidis</div> <em>The ISME Journal</em> (2021-12-09) <a href="https://doi.org/gpdvs9">https://doi.org/gpdvs9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41396-021-01149-9">10.1038/s41396-021-01149-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34887548">34887548</a></div></div>
</div>
<div id="ref-TzhWL2dj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">10. </div><div class="csl-right-inline"><strong>Ten years of pan-genome analyses</strong> <div class="csl-block">George Vernikos, Duccio Medini, David R Riley, Hervé Tettelin</div> <em>Current Opinion in Microbiology</em> (2015-02) <a href="https://doi.org/f63bmf">https://doi.org/f63bmf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.mib.2014.11.016">10.1016/j.mib.2014.11.016</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25483351">25483351</a></div></div>
</div>
<div id="ref-nSWbl26Q" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">11. </div><div class="csl-right-inline"><strong>Multiple levels of the unknown in microbiome research</strong> <div class="csl-block">Andrew Maltez Thomas, Nicola Segata</div> <em>BMC Biology</em> (2019-12) <a href="https://doi.org/gnm4t7">https://doi.org/gnm4t7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12915-019-0667-z">10.1186/s12915-019-0667-z</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31189463">31189463</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6560723">PMC6560723</a></div></div>
</div>
<div id="ref-6Fmlbtl9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">12. </div><div class="csl-right-inline"><strong>Metagenome-assembled genome binning methods with short reads disproportionately fail for plasmids and genomic Islands</strong> <div class="csl-block">Finlay Maguire, Baofeng Jia, Kristen L Gray, Wing Yin Venus Lau, Robert G Beiko, Fiona SL Brinkman</div> <em>Microbial Genomics</em> (2020-10-01) <a href="https://doi.org/gns4sc">https://doi.org/gns4sc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1099/mgen.0.000436">10.1099/mgen.0.000436</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33001022">33001022</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7660262">PMC7660262</a></div></div>
</div>
<div id="ref-1HDOysM7N" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">13. </div><div class="csl-right-inline"><strong>Toward quantifying the adaptive role of bacterial pangenomes during environmental perturbations</strong> <div class="csl-block">Roth E Conrad, Tomeu Viver, Juan F Gago, Janet K Hatt, Fanus Venter, Ramon Rosselló-Móra, Konstantinos T Konstantinidis</div> <em>Microbiology</em> (2021-03-15) <a href="https://doi.org/gns4sd">https://doi.org/gns4sd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2021.03.15.435471">10.1101/2021.03.15.435471</a></div></div>
</div>
<div id="ref-mAjlUXL2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">14. </div><div class="csl-right-inline"><strong>MetaPalette: a &lt;i&gt;k&lt;/i&gt; -mer Painting Approach for Metagenomic Taxonomic Profiling and Quantification of Novel Strain Variation</strong> <div class="csl-block">David Koslicki, Daniel Falush</div> <em>mSystems</em> (2016-06-28) <a href="https://doi.org/gg3gbd">https://doi.org/gg3gbd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1128/msystems.00020-16">10.1128/msystems.00020-16</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27822531">27822531</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5069763">PMC5069763</a></div></div>
</div>
<div id="ref-mH9pzoIn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">15. </div><div class="csl-right-inline"><strong>Mash: fast genome and metagenome distance estimation using MinHash</strong> <div class="csl-block">Brian D Ondov, Todd J Treangen, Páll Melsted, Adam B Mallonee, Nicholas H Bergman, Sergey Koren, Adam M Phillippy</div> <em>Genome Biology</em> (2016-12) <a href="https://doi.org/gfx74q">https://doi.org/gfx74q</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s13059-016-0997-x">10.1186/s13059-016-0997-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27323842">27323842</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4915045">PMC4915045</a></div></div>
</div>
<div id="ref-IadOz4ol" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">16. </div><div class="csl-right-inline"><strong>Lightweight compositional analysis of metagenomes with FracMinHash and minimum metagenome covers</strong> <div class="csl-block">Luiz Irber, Phillip T Brooks, Taylor Reiter, NTessa Pierce-Ward, Mahmudur Rahman Hera, David Koslicki, CTitus Brown</div> <em>Bioinformatics</em> (2022-01-12) <a href="https://doi.org/gn34zt">https://doi.org/gn34zt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2022.01.11.475838">10.1101/2022.01.11.475838</a></div></div>
</div>
<div id="ref-jroubkzG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">17. </div><div class="csl-right-inline"><strong>Genome-resolved metagenomics identifies genetic mobility, metabolic interactions, and unexpected diversity in perchlorate-reducing communities</strong> <div class="csl-block">Tyler P Barnum, Israel A Figueroa, Charlotte I Carlström, Lauren N Lucas, Anna L Engelbrektson, John D Coates</div> <em>The ISME Journal</em> (2018-06) <a href="https://doi.org/gdms93">https://doi.org/gdms93</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41396-018-0081-5">10.1038/s41396-018-0081-5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29476141">29476141</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5955982">PMC5955982</a></div></div>
</div>
<div id="ref-VsLophuT" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">18. </div><div class="csl-right-inline"><strong>Exploring neighborhoods in large metagenome assembly graphs using spacegraphcats reveals hidden sequence diversity</strong> <div class="csl-block">CTitus Brown, Dominik Moritz, Michael P O’Brien, Felix Reidl, Taylor Reiter, Blair D Sullivan</div> <em>Genome Biology</em> (2020-12) <a href="https://doi.org/d4bb">https://doi.org/d4bb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s13059-020-02066-4">10.1186/s13059-020-02066-4</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32631445">32631445</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7336657">PMC7336657</a></div></div>
</div>
<div id="ref-lsbnKJf8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">19. </div><div class="csl-right-inline"><strong>Critical Assessment of Metagenome Interpretation—a benchmark of metagenomics software</strong> <div class="csl-block">Alexander Sczyrba, Peter Hofmann, Peter Belmann, David Koslicki, Stefan Janssen, Johannes Dröge, Ivan Gregor, Stephan Majda, Jessika Fiedler, Eik Dahms, … Alice C McHardy</div> <em>Nature Methods</em> (2017-11) <a href="https://doi.org/gbzspt">https://doi.org/gbzspt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nmeth.4458">10.1038/nmeth.4458</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28967888">28967888</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5903868">PMC5903868</a></div></div>
</div>
<div id="ref-A7OCZbgE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">20. </div><div class="csl-right-inline"><strong>Critical Assessment of Metagenome Interpretation - the second round of challenges</strong> <div class="csl-block">F Meyer, A Fritz, Z-L Deng, D Koslicki, A Gurevich, G Robertson, M Alser, D Antipov, F Beghini, D Bertrand, … AC McHardy</div> <em>Bioinformatics</em> (2021-07-12) <a href="https://doi.org/gk566x">https://doi.org/gk566x</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2021.07.12.451567">10.1101/2021.07.12.451567</a></div></div>
</div>
<div id="ref-lQHniSOY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">21. </div><div class="csl-right-inline"><strong>Generation of lineage-resolved complete metagenome-assembled genomes by precision phasing</strong> <div class="csl-block">Derek M Bickhart, Mikhail Kolmogorov, Elizabeth Tseng, Daniel M Portik, Anton Korobeynikov, Ivan Tolstoganov, Gherman Uritskiy, Ivan Liachko, Shawn T Sullivan, Sung Bong Shin, … Timothy PL Smith</div> <em>Microbiology</em> (2021-05-04) <a href="https://doi.org/gns4sf">https://doi.org/gns4sf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2021.05.04.442591">10.1101/2021.05.04.442591</a></div></div>
</div>
<div id="ref-TO2yVgRA" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">22. </div><div class="csl-right-inline"><strong>Single-cell transcriptomics for the 99.9% of species without reference genomes</strong> <div class="csl-block">Olga Borisovna Botvinnik, Venkata Naga Pranathi Vemuri, NTessa Pierce, Phoenix Aja Logan, Saba Nafees, Lekha Karanam, Kyle Joseph Travaglini, Camille Sophie Ezran, Lili Ren, Yanyi Juang, … CTitus Brown</div> <em>Bioinformatics</em> (2021-07-10) <a href="https://doi.org/gns4sg">https://doi.org/gns4sg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2021.07.09.450799">10.1101/2021.07.09.450799</a></div></div>
</div>
<div id="ref-vFApiGX2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">23. </div><div class="csl-right-inline"><strong>GTDB: an ongoing census of bacterial and archaeal diversity through a phylogenetically consistent, rank normalized and complete genome-based taxonomy</strong> <div class="csl-block">Donovan H Parks, Maria Chuvochina, Christian Rinke, Aaron J Mussig, Pierre-Alain Chaumeil, Philip Hugenholtz</div> <em>Nucleic Acids Research</em> (2022-01-07) <a href="https://doi.org/gm97d8">https://doi.org/gm97d8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab776">10.1093/nar/gkab776</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34520557">34520557</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8728215">PMC8728215</a></div></div>
</div>
<div id="ref-MSOGoMuF" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">24. </div><div class="csl-right-inline"><strong>Comparative genomics: the bacterial pan-genome</strong> <div class="csl-block">Hervé Tettelin, David Riley, Ciro Cattuto, Duccio Medini</div> <em>Current Opinion in Microbiology</em> (2008-10) <a href="https://doi.org/dp79f2">https://doi.org/dp79f2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.mib.2008.09.006">10.1016/j.mib.2008.09.006</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19086349">19086349</a></div></div>
</div>
<div id="ref-b3TnpDgh" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">25. </div><div class="csl-right-inline"><strong>NCBI prokaryotic genome annotation pipeline</strong> <div class="csl-block">Tatiana Tatusova, Michael DiCuccio, Azat Badretdin, Vyacheslav Chetvernin, Eric P Nawrocki, Leonid Zaslavsky, Alexandre Lomsadze, Kim D Pruitt, Mark Borodovsky, James Ostell</div> <em>Nucleic Acids Research</em> (2016-08-19) <a href="https://doi.org/f82gsk">https://doi.org/f82gsk</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkw569">10.1093/nar/gkw569</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27342282">27342282</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5001611">PMC5001611</a></div></div>
</div>
<div id="ref-UUdrBbKz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">26. </div><div class="csl-right-inline"><strong>RefSeq: expanding the Prokaryotic Genome Annotation Pipeline reach with protein family model curation</strong> <div class="csl-block">Wenjun Li, Kathleen R O’Neill, Daniel H Haft, Michael DiCuccio, Vyacheslav Chetvernin, Azat Badretdin, George Coulouris, Farideh Chitsaz, Myra K Derbyshire, AScott Durkin, … Françoise Thibaud-Nissen</div> <em>Nucleic Acids Research</em> (2021-01-08) <a href="https://doi.org/gnrhsn">https://doi.org/gnrhsn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkaa1105">10.1093/nar/gkaa1105</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33270901">33270901</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7779008">PMC7779008</a></div></div>
</div>
<div id="ref-UotiKLjf" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">27. </div><div class="csl-right-inline"><strong>The Integrative Human Microbiome Project</strong> <div class="csl-block">The Integrative HMP (iHMP) Research Network Consortium</div> <em>Nature</em> (2019-05) <a href="https://doi.org/gf3wp9">https://doi.org/gf3wp9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41586-019-1238-8">10.1038/s41586-019-1238-8</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31142853">31142853</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6784865">PMC6784865</a></div></div>
</div>
<div id="ref-12fAf7MKk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">28. </div><div class="csl-right-inline"><strong>High throughput ANI analysis of 90K prokaryotic genomes reveals clear species boundaries</strong> <div class="csl-block">Chirag Jain, Luis M Rodriguez-R, Adam M Phillippy, Konstantinos T Konstantinidis, Srinivas Aluru</div> <em>Nature Communications</em> (2018-12) <a href="https://doi.org/gfknmg">https://doi.org/gfknmg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-018-07641-9">10.1038/s41467-018-07641-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30504855">30504855</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6269478">PMC6269478</a></div></div>
</div>
<div id="ref-JrcoonJ8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">29. </div><div class="csl-right-inline"><strong>Prokka: rapid prokaryotic genome annotation</strong> <div class="csl-block">T Seemann</div> <em>Bioinformatics</em> (2014-07-15) <a href="https://doi.org/f6b3k4">https://doi.org/f6b3k4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btu153">10.1093/bioinformatics/btu153</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24642063">24642063</a></div></div>
</div>
<div id="ref-1cFZOigd" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">30. </div><div class="csl-right-inline"><strong>Roary: rapid large-scale prokaryote pan genome analysis</strong> <div class="csl-block">Andrew J Page, Carla A Cummins, Martin Hunt, Vanessa K Wong, Sandra Reuter, Matthew TG Holden, Maria Fookes, Daniel Falush, Jacqueline A Keane, Julian Parkhill</div> <em>Bioinformatics</em> (2015-11-15) <a href="https://doi.org/ggbhmt">https://doi.org/ggbhmt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btv421">10.1093/bioinformatics/btv421</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26198102">26198102</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4817141">PMC4817141</a></div></div>
</div>
<div id="ref-ExvjTLL3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">31. </div><div class="csl-right-inline"><strong>sourmash: a library for MinHash sketching of DNA</strong> <div class="csl-block">C Titus Brown, Luiz Irber</div> <em>The Journal of Open Source Software</em> (2016-09-14) <a href="https://doi.org/ghdrk5">https://doi.org/ghdrk5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.00027">10.21105/joss.00027</a></div></div>
</div>
<div id="ref-6ZxBxq9m" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">32. </div><div class="csl-right-inline"><strong>VEGAN, a package of R functions for community ecology</strong> <div class="csl-block">Philip Dixon</div> <em>Journal of Vegetation Science</em> (2003-12) <a href="https://doi.org/dz9txb">https://doi.org/dz9txb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/j.1654-1103.2003.tb02228.x">10.1111/j.1654-1103.2003.tb02228.x</a></div></div>
</div>
<div id="ref-10u9VaxL4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">33. </div><div class="csl-right-inline"><strong>An object-oriented framework for evolutionary pangenome analysis</strong> <div class="csl-block">Ignacio Ferrés, Gregorio Iraola</div> <em>Cell Reports Methods</em> (2021-09) <a href="https://doi.org/gm3zts">https://doi.org/gm3zts</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.crmeth.2021.100085">10.1016/j.crmeth.2021.100085</a></div></div>
</div>
<div id="ref-wS8KrahX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">34. </div><div class="csl-right-inline"><strong>Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification</strong> <div class="csl-block">Oliver Schwengers, Lukas Jelonek, Marius Alfred Dieckmann, Sebastian Beyvers, Jochen Blom, Alexander Goesmann</div> <em>Microbial Genomics</em> (2021-11-05) <a href="https://doi.org/gnfrj7">https://doi.org/gnfrj7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1099/mgen.0.000685">10.1099/mgen.0.000685</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34739369">34739369</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8743544">PMC8743544</a></div></div>
</div>
<div id="ref-LVpd95E4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">35. </div><div class="csl-right-inline"><strong>&lt;i&gt;Polyester&lt;/i&gt; : simulating RNA-seq datasets with differential transcript expression</strong> <div class="csl-block">Alyssa C Frazee, Andrew E Jaffe, Ben Langmead, Jeffrey T Leek</div> <em>Bioinformatics</em> (2015-09-01) <a href="https://doi.org/f7rwtr">https://doi.org/f7rwtr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btv272">10.1093/bioinformatics/btv272</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25926345">25926345</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4635655">PMC4635655</a></div></div>
</div>
<div id="ref-19Z0U0zL4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">36. </div><div class="csl-right-inline"><strong>Multi-omics of the gut microbial ecosystem in inflammatory bowel diseases</strong> <div class="csl-block">IBDMDB Investigators, Jason Lloyd-Price, Cesar Arze, Ashwin N Ananthakrishnan, Melanie Schirmer, Julian Avila-Pacheco, Tiffany W Poon, Elizabeth Andrews, Nadim J Ajami, Kevin S Bonham, … Curtis Huttenhower</div> <em>Nature</em> (2019-05) <a href="https://doi.org/ggd6wc">https://doi.org/ggd6wc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41586-019-1237-9">10.1038/s41586-019-1237-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31142855">31142855</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6650278">PMC6650278</a></div></div>
</div>
<div id="ref-G2CzfHHF" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">37. </div><div class="csl-right-inline"><strong>The khmer software package: enabling efficient nucleotide sequence analysis</strong> <div class="csl-block">Michael R Crusoe, Hussien F Alameldin, Sherine Awad, Elmar Boucher, Adam Caldwell, Reed Cartwright, Amanda Charbonneau, Bede Constantinides, Greg Edvenson, Scott Fay, … CTitus Brown</div> <em>F1000Research</em> (2015-09-25) <a href="https://doi.org/9qp">https://doi.org/9qp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.12688/f1000research.6924.1">10.12688/f1000research.6924.1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26535114">26535114</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4608353">PMC4608353</a></div></div>
</div>
</div>
<h1 id="appendixsupplementary-information">Appendix/Supplementary information</h1>
<div id="fig:tree_fig" class="fignos">
<figure>
<img src="images/tree_fig.png" alt="Figure 8: Organisms used in this paper." /><figcaption aria-hidden="true"><span>Figure 8:</span> <strong>Organisms used in this paper.</strong></figcaption>
</figure>
</div>
<div id="fig:orpheum_db" class="fignos">
<figure>
<img src="images/orpheum_supp1.png" style="height:3in" alt="Figure 9: The slight increase observable for some species is a results in different thresholds, where we used 0.39 for the species database and 0.5 for the GTDB rs202 database." /><figcaption aria-hidden="true"><span>Figure 9:</span> The slight increase observable for some species is a results in different thresholds, where we used 0.39 for the species database and 0.5 for the GTDB rs202 database.</figcaption>
</figure>
</div>
<div id="tbl:gtdb_acc" class="tablenos">
<table id="tbl:gtdb_acc">
<caption><span>Table 4:</span> GTDB genomes used to benchmark orpheum accuracy.
</caption>
<colgroup>
<col style="width: 6%" />
<col style="width: 5%" />
<col style="width: 8%" />
<col style="width: 9%" />
<col style="width: 8%" />
<col style="width: 10%" />
<col style="width: 9%" />
<col style="width: 14%" />
<col style="width: 4%" />
<col style="width: 22%" />
</colgroup>
<thead>
<tr class="header">
<th>accession</th>
<th>superkingdom</th>
<th>phylum</th>
<th>class</th>
<th>order</th>
<th>family</th>
<th>genus</th>
<th>species</th>
<th>NCBI taxid</th>
<th>NCBI organism name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GCF_000012125.1</td>
<td>d__Bacteria</td>
<td>p__Chlamydiota</td>
<td>c__Chlamydiia</td>
<td>o__Chlamydiales</td>
<td>f__Chlamydiaceae</td>
<td>g__Chlamydia</td>
<td>s__Chlamydia trachomatis</td>
<td>315277</td>
<td>Chlamydia trachomatis A/HAR-13</td>
</tr>
<tr class="even">
<td>GCF_000381045.1</td>
<td>d__Bacteria</td>
<td>p__Deinococcota</td>
<td>c__Deinococci</td>
<td>o__Deinococcales</td>
<td>f__Thermaceae</td>
<td>g__Thermus</td>
<td>s__Thermus scotoductus</td>
<td>1123391</td>
<td>Thermus scotoductus DSM 8553</td>
</tr>
<tr class="odd">
<td>GCF_900156205.1</td>
<td>d__Bacteria</td>
<td>p__Spirochaetota</td>
<td>c__Leptospirae</td>
<td>o__Leptospirales</td>
<td>f__Leptospiraceae</td>
<td>g__Leptospira</td>
<td>s__Leptospira interrogans</td>
<td>173</td>
<td>Leptospira interrogans</td>
</tr>
<tr class="even">
<td>GCA_900315925.1</td>
<td>d__Bacteria</td>
<td>p__Firmicutes_C</td>
<td>c__Negativicutes</td>
<td>o__Acidaminococcales</td>
<td>f__Acidaminococcaceae</td>
<td>g__Succiniclasticum</td>
<td>s__Succiniclasticum sp900315925</td>
<td>1387507</td>
<td>uncultured Selenomonadales bacterium</td>
</tr>
<tr class="odd">
<td>GCF_004006635.1</td>
<td>d__Bacteria</td>
<td>p__Fusobacteriota</td>
<td>c__Fusobacteriia</td>
<td>o__Fusobacteriales</td>
<td>f__Fusobacteriaceae</td>
<td>g__Fusobacterium_C</td>
<td>s__Fusobacterium_C necrophorum</td>
<td>143388</td>
<td>Fusobacterium necrophorum subsp. necrophorum</td>
</tr>
<tr class="even">
<td>GCA_903857495.1</td>
<td>d__Bacteria</td>
<td>p__Acidobacteriota</td>
<td>c__Holophagae</td>
<td>o__Holophagales</td>
<td>f__Holophagaceae</td>
<td>g__Geothrix</td>
<td>s__Geothrix sp903857495</td>
<td>904990</td>
<td>uncultured Holophagaceae bacterium</td>
</tr>
<tr class="odd">
<td>GCF_006742205.1</td>
<td>d__Bacteria</td>
<td>p__Firmicutes</td>
<td>c__Bacilli</td>
<td>o__Staphylococcales</td>
<td>f__Staphylococcaceae</td>
<td>g__Staphylococcus</td>
<td>s__Staphylococcus epidermidis</td>
<td>1282</td>
<td>Staphylococcus epidermidis</td>
</tr>
<tr class="even">
<td>GCF_002006445.1</td>
<td>d__Bacteria</td>
<td>p__Firmicutes_A</td>
<td>c__Clostridia</td>
<td>o__Clostridiales</td>
<td>f__Clostridiaceae</td>
<td>g__Clostridium</td>
<td>s__Clostridium beijerinckii</td>
<td>1520</td>
<td>Clostridium beijerinckii</td>
</tr>
<tr class="odd">
<td>GCA_903861715.1</td>
<td>d__Bacteria</td>
<td>p__Patescibacteria</td>
<td>c__Paceibacteria</td>
<td>o__Moranbacterales</td>
<td>f__GWC2-37-73</td>
<td>g__CAIPMZ01</td>
<td>s__CAIPMZ01 sp903861715</td>
<td>77133</td>
<td>uncultured bacterium</td>
</tr>
<tr class="even">
<td>GCF_000008025.1</td>
<td>d__Bacteria</td>
<td>p__Proteobacteria</td>
<td>c__Alphaproteobacteria</td>
<td>o__Rickettsiales</td>
<td>f__Anaplasmataceae</td>
<td>g__Wolbachia</td>
<td>s__Wolbachia pipientis</td>
<td>163164</td>
<td>Wolbachia endosymbiont of Drosophila melanogaster</td>
</tr>
<tr class="odd">
<td>GCF_000830885.1</td>
<td>d__Bacteria</td>
<td>p__Chloroflexota</td>
<td>c__Dehalococcoidia</td>
<td>o__Dehalococcoidales</td>
<td>f__Dehalococcoidaceae</td>
<td>g__Dehalococcoides</td>
<td>s__Dehalococcoides mccartyi_B</td>
<td>1432061</td>
<td>Dehalococcoides mccartyi CG5</td>
</tr>
<tr class="even">
<td>GCF_000299235.1</td>
<td>d__Bacteria</td>
<td>p__Nitrospirota_A</td>
<td>c__Leptospirillia</td>
<td>o__Leptospirillales</td>
<td>f__Leptospirillaceae</td>
<td>g__Leptospirillum_A</td>
<td>s__Leptospirillum_A rubarum</td>
<td>1048260</td>
<td>Leptospirillum ferriphilum ML-04</td>
</tr>
<tr class="odd">
<td>GCA_000635495.1</td>
<td>d__Bacteria</td>
<td>p__Cyanobacteria</td>
<td>c__Cyanobacteriia</td>
<td>o__PCC-6307</td>
<td>f__Cyanobiaceae</td>
<td>g__Prochlorococcus_A</td>
<td>s__Prochlorococcus_A sp000635495</td>
<td>1471472</td>
<td>Prochlorococcus sp. scB243_495K23</td>
</tr>
<tr class="even">
<td>GCA_903852495.1</td>
<td>d__Bacteria</td>
<td>p__Desulfobacterota_F</td>
<td>c__Desulfuromonadia</td>
<td>o__Geobacterales</td>
<td>f__Pseudopelobacteraceae</td>
<td>g__Pseudopelobacter</td>
<td>s__Pseudopelobacter sp903852495</td>
<td>214033</td>
<td>uncultured Geobacteraceae bacterium</td>
</tr>
<tr class="odd">
<td>GCA_903869265.1</td>
<td>d__Bacteria</td>
<td>p__Desulfobacterota</td>
<td>c__Desulfobulbia</td>
<td>o__Desulfobulbales</td>
<td>f__Desulfurivibrionaceae</td>
<td>g__UBA2262</td>
<td>s__UBA2262 sp903869265</td>
<td>34034</td>
<td>uncultured delta proteobacterium</td>
</tr>
<tr class="even">
<td>GCF_000020225.1</td>
<td>d__Bacteria</td>
<td>p__Verrucomicrobiota</td>
<td>c__Verrucomicrobiae</td>
<td>o__Verrucomicrobiales</td>
<td>f__Akkermansiaceae</td>
<td>g__Akkermansia</td>
<td>s__Akkermansia muciniphila</td>
<td>349741</td>
<td>Akkermansia muciniphila ATCC BAA-835</td>
</tr>
<tr class="odd">
<td>GCF_900478295.1</td>
<td>d__Bacteria</td>
<td>p__Campylobacterota</td>
<td>c__Campylobacteria</td>
<td>o__Campylobacterales</td>
<td>f__Helicobacteraceae</td>
<td>g__Helicobacter</td>
<td>s__Helicobacter pylori</td>
<td>102618</td>
<td>Helicobacter pylori NCTC 11637 = CCUG 17874 = ATCC 43504</td>
</tr>
<tr class="even">
<td>GCF_000154205.1</td>
<td>d__Bacteria</td>
<td>p__Bacteroidota</td>
<td>c__Bacteroidia</td>
<td>o__Bacteroidales</td>
<td>f__Bacteroidaceae</td>
<td>g__Bacteroides</td>
<td>s__Bacteroides uniformis</td>
<td>411479</td>
<td>Bacteroides uniformis ATCC 8492</td>
</tr>
<tr class="odd">
<td>GCF_000196555.1</td>
<td>d__Bacteria</td>
<td>p__Actinobacteriota</td>
<td>c__Actinomycetia</td>
<td>o__Actinomycetales</td>
<td>f__Bifidobacteriaceae</td>
<td>g__Bifidobacterium</td>
<td>s__Bifidobacterium longum</td>
<td>565042</td>
<td>Bifidobacterium longum subsp. longum JCM 1217</td>
</tr>
<tr class="even">
<td>GCF_000012285.1</td>
<td>d__Archaea</td>
<td>p__Thermoproteota</td>
<td>c__Thermoproteia</td>
<td>o__Sulfolobales</td>
<td>f__Sulfolobaceae</td>
<td>g__Sulfolobus</td>
<td>s__Sulfolobus acidocaldarius</td>
<td>330779</td>
<td>Sulfolobus acidocaldarius DSM 639</td>
</tr>
<tr class="odd">
<td>GCF_000970205.1</td>
<td>d__Archaea</td>
<td>p__Halobacteriota</td>
<td>c__Methanosarcinia</td>
<td>o__Methanosarcinales</td>
<td>f__Methanosarcinaceae</td>
<td>g__Methanosarcina</td>
<td>s__Methanosarcina mazei</td>
<td>213585</td>
<td>Methanosarcina mazei S-6</td>
</tr>
<tr class="even">
<td>GCA_900314695.1</td>
<td>d__Archaea</td>
<td>p__Methanobacteriota</td>
<td>c__Methanobacteria</td>
<td>o__Methanobacteriales</td>
<td>f__Methanobacteriaceae</td>
<td>g__Methanobrevibacter_A</td>
<td>s__Methanobrevibacter_A sp900314695</td>
<td>253161</td>
<td>uncultured Methanobrevibacter sp.</td>
</tr>
<tr class="odd">
<td>GCF_900083515.1</td>
<td>d__Archaea</td>
<td>p__Thermoplasmatota</td>
<td>c__Thermoplasmata</td>
<td>o__Thermoplasmatales</td>
<td>f__Thermoplasmataceae</td>
<td>g__Cuniculiplasma</td>
<td>s__Cuniculiplasma divulgatum</td>
<td>1673428</td>
<td>Cuniculiplasma divulgatum</td>
</tr>
</tbody>
</table>
</div>
<div id="tbl:refseq_acc" class="tablenos">
<table id="tbl:refseq_acc">
<caption><span>Table 5:</span> RefSeq genomes not in the GTDB rs202 database used to benchmark orpheum accuracy.
</caption>
<thead>
<tr class="header">
<th>accession</th>
<th>NCBI taxid</th>
<th>NCBI organism name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GCF_003428625.2</td>
<td>2303751</td>
<td>Acidipila sp. 4G-K13</td>
</tr>
<tr class="even">
<td>GCF_001700755.2</td>
<td>1160719</td>
<td>Cutibacterium granulosum DSM 20700</td>
</tr>
<tr class="odd">
<td>GCF_001884725.2</td>
<td>336810</td>
<td>Candidatus Sulcia muelleri</td>
</tr>
<tr class="even">
<td>GCF_015356815.2</td>
<td>225148</td>
<td>Candidatus Rhabdochlamydia porcellionis</td>
</tr>
<tr class="odd">
<td>GCF_019599295.1</td>
<td>2866714</td>
<td>Oscillochloris sp. ZM17-4</td>
</tr>
<tr class="even">
<td>GCF_020520145.1</td>
<td>936456</td>
<td>Desulfurispirillum indicum</td>
</tr>
<tr class="odd">
<td>GCF_019175305.1</td>
<td>2286</td>
<td>Saccharolobus shibatae</td>
</tr>
<tr class="even">
<td>GCF_018282115.1</td>
<td>2732530</td>
<td>Synechocystis sp. PCC 7338</td>
</tr>
<tr class="odd">
<td>GCF_018863415.1</td>
<td>2774531</td>
<td>Deinococcus sp. SYSU M49105</td>
</tr>
<tr class="even">
<td>GCF_013456555.2</td>
<td>1710539</td>
<td>Natrinema sp. YPL30</td>
</tr>
<tr class="odd">
<td>GCF_000167435.2</td>
<td>1314</td>
<td>Streptococcus pyogenes</td>
</tr>
<tr class="even">
<td>GCF_018205295.1</td>
<td>859</td>
<td>Fusobacterium necrophorum</td>
</tr>
<tr class="odd">
<td>GCF_019173545.1</td>
<td>1455061</td>
<td>Candidatus Magnetobacterium casensis</td>
</tr>
<tr class="even">
<td>GCF_018398935.1</td>
<td>1123043</td>
<td>Telmatocola sphagniphila</td>
</tr>
<tr class="odd">
<td>GCF_000145825.2</td>
<td>629264</td>
<td>Pseudomonas syringae Cit 7</td>
</tr>
<tr class="even">
<td>GCF_002442595.2</td>
<td>139</td>
<td>Borreliella burgdorferi</td>
</tr>
<tr class="odd">
<td>GCF_009156025.2</td>
<td>28903</td>
<td>Mycoplasmopsis bovis</td>
</tr>
<tr class="even">
<td>GCF_019688735.1</td>
<td>2867247</td>
<td>Thermosulfurimonas sp. F29</td>
</tr>
<tr class="odd">
<td>GCF_018588215.1</td>
<td>1755816</td>
<td>Thermosipho sp. 1244</td>
</tr>
<tr class="even">
<td>GCF_018336995.1</td>
<td>239935</td>
<td>Akkermansia muciniphila</td>
</tr>
</tbody>
</table>
</div>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 12pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    word-break: break-word;
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
